# Foreword
Los archivos a obtener son:
* [Elecciones presidenciales, votación en Chile, 1989 a 2017](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_presidenciales_ce_1989_2017_Chile.xlsx)
* [Elecciones presidenciales, votación en el extranjero, 2017](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_presidenciales_ce_2017_Extranjero.xlsx)
* [Elecciones de senadores, 1989 a 2017](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_senadores_ce_1989_2017.xlsx)
* [Elecciones de diputados, 1989 a 2017](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_diputados_ce_1989_2017.xlsx)
* [Elecciones de alcaldes, 2004 a 2016](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_alcaldes_ce_2004_2016.xlsx)
* [Elecciones de concejales, 2004 a 2016](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_concejales_ce_2004_2016.xlsx)
* [Elecciones de alcaldes y concejales, 1992 a 2000](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_concejales_ce_1992_al_2000.xlsx)
* [Consejeros regionales, 2013 a 2017](https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_core_ce_2013_2017.xlsx)

Todos estos archivos fueron obtenidos desde el sitio oficial de [Servel](https://www.servel.cl), particularmente de la sección [Resultados en Excel por circunscripción electoral](https://www.servel.cl/elecciones-primarias-2013-por-circunscripcion-electoral/). 

In [1]:
pwd

'/mnt/c/Users/javie/Documents/Projects/Analytics/20191116-elecciones_chile'

In [2]:
pip install wget

Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install xlrd

Note: you may need to restart the kernel to use updated packages.


In [4]:
import wget
import subprocess
import pandas as pd
import numpy as np
import matplotlib as mpl

In [5]:
wget.download(url = "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_presidenciales_ce_1989_2017_Chile.xlsx",
              out = "presidenciales_test.xlsx")

'presidenciales_test.xlsx'

In [6]:
%%bash
ls -lA

total 102316
drwxrwxrwx 1 jtapia jtapia     4096 Nov 16 20:29 .ipynb_checkpoints
-rwxrwxrwx 1 jtapia jtapia    61591 Dec 16 14:11 20191116-etl_elecciones.ipynb
-rwxrwxrwx 1 jtapia jtapia  2190334 Dec 16 14:07 alcaldes-2004_2016.xlsx
-rwxrwxrwx 1 jtapia jtapia  5269614 Dec 16 14:08 alcaldes_concejales-1992_2000.xlsx
-rwxrwxrwx 1 jtapia jtapia 12273597 Dec 16 14:08 concejales-2004_2016.xlsx
-rwxrwxrwx 1 jtapia jtapia  3940352 Nov 26 00:42 concejales-2004_2016.xlsxfq_1b5b0.tmp
-rwxrwxrwx 1 jtapia jtapia  3877456 Dec 16 14:08 cores-2003_2017.xlsx
-rwxrwxrwx 1 jtapia jtapia  8897483 Dec 16 14:07 diputados-1989_2017.xlsx
-rwxrwxrwx 1 jtapia jtapia  6384518 Dec 16 14:07 presidenciales-1989_2017.xlsx
-rwxrwxrwx 1 jtapia jtapia   127582 Dec 16 14:07 presidenciales-2017-ext.xlsx
-rwxrwxrwx 1 jtapia jtapia  6384518 Dec 16 14:14 presidenciales_test.xlsx
-rwxrwxrwx 1 jtapia jtapia 51262553 Nov 25 17:23 resultados.csv
-rwxrwxrwx 1 jtapia jtapia  4077460 Dec 16 14:07 senadores-1989_2017.xlsx


In [7]:
%%bash
find . -name "*xlsx" -exec rm -f {} \;

In [8]:
to_download = ["https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_presidenciales_ce_1989_2017_Chile.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_presidenciales_ce_2017_Extranjero.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_senadores_ce_1989_2017.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_diputados_ce_1989_2017.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_alcaldes_ce_2004_2016.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_concejales_ce_2004_2016.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_concejales_ce_1992_al_2000.xlsx",
               "https://www.servel.cl/wp-content/uploads/2019/04/resultados_elecciones_core_ce_2013_2017.xlsx"]

output_names = ["presidenciales-1989_2017.xlsx",
                "presidenciales-2017-ext.xlsx",
                "senadores-1989_2017.xlsx",
                "diputados-1989_2017.xlsx",
                "alcaldes-2004_2016.xlsx",
                "concejales-2004_2016.xlsx",
                "alcaldes_concejales-1992_2000.xlsx",
                "cores-2003_2017.xlsx"]

download_dict = dict(zip(output_names, to_download))

variable_names = ["presidentes", "presidentes_ext", "senadores", "diputados", "alcaldes", 
                  "concejales", "alc_and_con", "cores"]

In [9]:
for output, link in download_dict.items():
#     print(link, " - ", output)
    wget.download(url = link, out = output)

In [10]:
ls_result = subprocess.check_output(["ls", "-l", "-A"])
ls_result.decode().split("\n")

['total 96076',
 'drwxrwxrwx 1 jtapia jtapia     4096 Nov 16 20:29 .ipynb_checkpoints',
 '-rwxrwxrwx 1 jtapia jtapia    58786 Dec 16 14:15 20191116-etl_elecciones.ipynb',
 '-rwxrwxrwx 1 jtapia jtapia  2190334 Dec 16 14:15 alcaldes-2004_2016.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia  5269614 Dec 16 14:15 alcaldes_concejales-1992_2000.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia 12273597 Dec 16 14:15 concejales-2004_2016.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia  3940352 Nov 26 00:42 concejales-2004_2016.xlsxfq_1b5b0.tmp',
 '-rwxrwxrwx 1 jtapia jtapia  3877456 Dec 16 14:15 cores-2003_2017.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia  8897483 Dec 16 14:15 diputados-1989_2017.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia  6384518 Dec 16 14:14 presidenciales-1989_2017.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia   127582 Dec 16 14:14 presidenciales-2017-ext.xlsx',
 '-rwxrwxrwx 1 jtapia jtapia 51262553 Nov 25 17:23 resultados.csv',
 '-rwxrwxrwx 1 jtapia jtapia  4077460 Dec 16 14:15 senadores-1989_2017.xlsx',
 '']

In [11]:
for i in range(0, len(variable_names)):
    vars()[variable_names[i]] = pd.read_excel(output_names[i])

In [12]:
del ls_result, output_names, to_download, output, link, i, download_dict

In [13]:
%whos

Variable          Type         Data/Info
----------------------------------------
alc_and_con       DataFrame          Tipo de Elección   <...>[41472 rows x 25 columns]
alcaldes          DataFrame          Tipo de Elección   <...>[18706 rows x 25 columns]
concejales        DataFrame          Tipo de Elección   <...>[98940 rows x 25 columns]
cores             DataFrame          Tipo de Elección   <...>[33770 rows x 25 columns]
diputados         DataFrame          Tipo de Elección   <...>[78995 rows x 25 columns]
mpl               module       <module 'matplotlib' from<...>/matplotlib/__init__.py'>
np                module       <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
pd                module       <module 'pandas' from '/h<...>ages/pandas/__init__.py'>
presidentes       DataFrame          Tipo de Elección   <...>[64366 rows x 24 columns]
presidentes_ext   DataFrame         Tipo de Elección    <...>n[1540 rows x 21 columns]
senadores         DataFrame          Tipo de Ele

In [14]:
tmp = []

for item in variable_names:
    tmp.append(vars()[item].columns.values.tolist())

tmp

[['Tipo de Elección',
  'Cargo',
  'Fecha de Elección',
  'Año de Elección',
  'Inicio de Período',
  'Fin de Período',
  'Período',
  'Votación Presidencial',
  'Id Región',
  'Región',
  'Provincia',
  'Nombre Provincia',
  'Comuna',
  'Circunscripción Electoral',
  'Sexo Mesa',
  'Candidato (a)',
  'Nombres',
  'Apellido Paterno',
  'Apellido Materno',
  'Sexo',
  'Electo(a)',
  'Partido',
  'Sigla Partido',
  'Votos Totales'],
 ['Tipo de Elección',
  'Cargo',
  'Fecha de Elección',
  'Año de Elección',
  'Inicio de Período',
  'Fin de Período',
  'Período',
  'Votación Presidencial',
  'Continente',
  'Pais',
  'Circunscripción Electoral',
  'Sexo Mesa',
  'Candidato (a)',
  'Nombres',
  'Apellido Paterno',
  'Apellido Materno',
  'Sexo',
  'Electo (a)',
  'Partido',
  'Sigla Partido',
  'Votos Totales'],
 ['Tipo de Elección',
  'Cargo',
  'Fecha de Elección',
  'Año de Elección',
  'Inicio de Período',
  'Fin de Período',
  'Período',
  'Id Región',
  'Región',
  'Circunscripción 

In [15]:
unique_data = set(x for list in tmp for x in list)
unique_data

{'Apellido Materno',
 'Apellido Paterno',
 'Año de Elección',
 'Candidato (a)',
 'Cargo',
 'Circunscripción Electoral',
 'Circunscripción Provincial',
 'Circunscripción Senatorial',
 'Comuna',
 'Continente',
 'Distrito',
 'Electo',
 'Electo (a)',
 'Electo(a)',
 'Fecha de Elección',
 'Fin de Período',
 'Id Región',
 'Inicio de Período',
 'Lista',
 'Nombre Circunscripción Electoral',
 'Nombre Provincia',
 'Nombres',
 'Pais',
 'Partido',
 'Período',
 'Provincia',
 'Región',
 'Sexo',
 'Sexo Mesa',
 'Sigla Partido',
 'Tipo de Elección',
 'Votación Presidencial',
 'Votos Totales'}

In [16]:
selected_vars = ["Candidato (a)", "Comuna", "Distrito", 
                 "Fecha de Elección", "Región", "Lista", "Nombre Provincia", 
                 "Nombre Circunscripción Electoral", "Partido", 
                 "Sexo", "Votación Presidencial", "Votos Totales",
                 "Cargo", "Electo", "Electo (a)", "Electo(a)",
                 "Tipo de Elección", "Período"]

In [17]:
base_tmp = presidentes.append(presidentes_ext).append(senadores).append(diputados).append(alcaldes). \
    append(concejales).append(alc_and_con).append(cores)

base_total = base_tmp.loc[:, base_tmp.columns.isin(selected_vars)]. \
    rename(columns = {"Candidato (a)": "candidato",
                      "Comuna": "comuna",
                      "Distrito": "distrito",
                      "Fecha de Elección": "fecha",
                      "Región": "region",
                      "Lista": "lista",
                      "Nombre Provincia": "provincia",
                      "Nombre Circunscripción Electoral": "circunscripcion",
                      "Partido": "partido",
                      "Sexo": "sexo",
                      "Votación Presidencial": "tipo_voto_presidencial",
                      "Votos Totales": "votos",
                      "Cargo": "cargo",
                      "Electo": "electo_t1",
                      "Electo(a)": "electo_t2",
                      "Electo (a)": "electo_t3",
                      "Tipo de Elección": "tipo_eleccion",
                      "Período": "periodo"
                     })

/home/jtapia/anaconda3/envs/data_analysis/lib/python3.6/site-packages/pandas/core/frame.py:7138: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort,


In [18]:
base_total.head()

candidato       cargo comuna  distrito electo_t1  \
0  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE  ARICA       NaN       NaN   
1                HERNAN BUCHI  PRESIDENTE  ARICA       NaN       NaN   
2             PATRICIO AYLWIN  PRESIDENTE  ARICA       NaN       NaN   
3                VOTOS NULOS   PRESIDENTE  ARICA       NaN       NaN   
4            VOTOS EN BLANCO   PRESIDENTE  ARICA       NaN       NaN   

  electo_t3 electo_t2      fecha lista circunscripcion provincia  \
0       NaN       NaN 1989-12-14   NaN             NaN     ARICA   
1       NaN       NaN 1989-12-14   NaN             NaN     ARICA   
2       NaN        SI 1989-12-14   NaN             NaN     ARICA   
3       NaN       NaN 1989-12-14   NaN             NaN     ARICA   
4       NaN       NaN 1989-12-14   NaN             NaN     ARICA   

               partido      periodo       region    sexo tipo_eleccion  \
0        INDEPENDIENTE  1990 a 1994  DE TARAPACA  HOMBRE  PRESIDENCIAL   
1        INDEPENDIENTE  1990 a 1994  DE TARAPACA  HOMBRE  PRESIDENCIAL   
2  DEMOCRATA CRISTIANO  1990 a 1994  DE TARAPACA  HOMBRE  PRESIDENCIAL   
3                  NaN  1990 a 1994  DE TARAPACA     NaN  PRESIDENCIAL   
4                  NaN  1990 a 1994  DE TARAPACA     NaN  PRESIDENCIAL   

  tipo_voto_presidencial  votos  
0         UNICA VOTACIÓN   7065  
1         UNICA VOTACIÓN  11549  
2         UNICA VOTACIÓN  23230  
3         UNICA VOTACIÓN    561  
4         UNICA VOTACIÓN    392

In [19]:
base_total.reset_index(inplace = True)

In [20]:
base_total.shape

(374720, 19)

El total de filas es de 374,720. Este número queda guardado como `total_records`, para futuras validaciones.

In [21]:
total_records = base_total.shape[0]
total_records

374720

In [22]:
del variable_names, presidentes, presidentes_ext, diputados, senadores, alcaldes, concejales, alc_and_con, cores, \
    base_tmp, item, selected_vars, unique_data

In [23]:
total_votes = base_total.votos.sum()
total_votes

254409627

El total de votos es de 254,409,627. Este número queda guardado como `total_votes`, para futuras validaciones.

In [24]:
%whos

Variable        Type         Data/Info
--------------------------------------
base_total      DataFrame            index            <...>374720 rows x 19 columns]
mpl             module       <module 'matplotlib' from<...>/matplotlib/__init__.py'>
np              module       <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
pd              module       <module 'pandas' from '/h<...>ages/pandas/__init__.py'>
subprocess      module       <module 'subprocess' from<...>python3.6/subprocess.py'>
tmp             list         n=8
total_records   int          374720
total_votes     int64        254409627
wget            module       <module 'wget' from '/hom<...>6/site-packages/wget.py'>


Lo primero que vamos a arreglar es la cantidad de columnas "Electo" (`Electo`, `Electo (a)` y `Electo(a)`). Se consolidarán las tres en una sola, que indique si el candidato fue ganador de la elección o no.

In [25]:
print(base_total.electo_t1.unique())
print(base_total.electo_t2.unique())
print(base_total.electo_t3.unique())

[nan 'SI' ' ']
[nan 'SI' '2ª Votación' ' ' 'CONCEJAL' '  ' 'ALCALDE']
[nan '2ª Votación' ' ' 'SI']


Viendo los datos que contienen, ahora sabemos que se perderá cierto nivel de información (el cargo para el cual fue electo el candidato, y si fue elección única o el candidato pasa a segunda vuelta). Quizás esa información se puede obtener con múltiples variables, como `electo` y `tipo_voto_presidencial`.

In [26]:
base_total.shape

(374720, 19)

In [27]:
base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.isnull())].shape

(205563, 19)

In [28]:
print(base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.notnull())].shape)
print(base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.notnull()) & (base_total.electo_t3.isnull())].shape)
print(base_total[(base_total.electo_t1.notnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.isnull())].shape)

(440, 19)
(87715, 19)
(81002, 19)


In [29]:
tmp_1 = base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.isnull())].shape[0]
tmp_2 = base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.notnull())].shape[0]
tmp_3 = base_total[(base_total.electo_t1.isnull()) & (base_total.electo_t2.notnull()) & (base_total.electo_t3.isnull())].shape[0]
tmp_4 = base_total[(base_total.electo_t1.notnull()) & (base_total.electo_t2.isnull()) & (base_total.electo_t3.isnull())].shape[0]

total_records == tmp_1 + tmp_2 + tmp_3 + tmp_4

True

In [30]:
del tmp_1, tmp_2, tmp_3, tmp_4

In [31]:
condiciones_electo = [
    (base_total["electo_t1"].isnull()) & (base_total["electo_t2"].isnull()) & (base_total["electo_t3"].isnull()),
    (base_total["electo_t1"].isnull()) & (base_total["electo_t2"].isnull()) & (base_total["electo_t3"].notnull()),
    (base_total["electo_t1"].isnull()) & (base_total["electo_t2"].notnull()) & (base_total["electo_t3"].isnull()),
    (base_total["electo_t1"].notnull()) & (base_total["electo_t2"].isnull()) & (base_total["electo_t3"].isnull())
]

resultados_electo = ["no", "si", "si", "si"]

base_total["electo"] = np.select(condiciones_electo, resultados_electo, default = "no")

base_total.drop(columns = ["electo_t1", "electo_t2", "electo_t3"], inplace = True)

In [32]:
base_total.head()

index                   candidato       cargo comuna  distrito      fecha  \
0      0  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE  ARICA       NaN 1989-12-14   
1      1                HERNAN BUCHI  PRESIDENTE  ARICA       NaN 1989-12-14   
2      2             PATRICIO AYLWIN  PRESIDENTE  ARICA       NaN 1989-12-14   
3      3                VOTOS NULOS   PRESIDENTE  ARICA       NaN 1989-12-14   
4      4            VOTOS EN BLANCO   PRESIDENTE  ARICA       NaN 1989-12-14   

  lista circunscripcion provincia              partido      periodo  \
0   NaN             NaN     ARICA        INDEPENDIENTE  1990 a 1994   
1   NaN             NaN     ARICA        INDEPENDIENTE  1990 a 1994   
2   NaN             NaN     ARICA  DEMOCRATA CRISTIANO  1990 a 1994   
3   NaN             NaN     ARICA                  NaN  1990 a 1994   
4   NaN             NaN     ARICA                  NaN  1990 a 1994   

        region    sexo tipo_eleccion tipo_voto_presidencial  votos electo  
0  DE TARAPACA  HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN   7065     no  
1  DE TARAPACA  HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  11549     no  
2  DE TARAPACA  HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  23230     si  
3  DE TARAPACA     NaN  PRESIDENCIAL         UNICA VOTACIÓN    561     no  
4  DE TARAPACA     NaN  PRESIDENCIAL         UNICA VOTACIÓN    392     no

In [33]:
print(base_total.shape)
print(base_total.electo.unique())

(374720, 17)
['no' 'si']


Existe ahora una única variable que indica si el candidato fue exitoso o no en la elección. Valido ahora que la información de segunda votación se pueda obtener con `electo`, `tipo_voto_presidencial` y `periodo`.

In [34]:
base_total.tipo_voto_presidencial.unique()

array(['UNICA VOTACIÓN', 'PRIMERA VOTACIÓN', 'SEGUNDA VOTACIÓN', nan],
      dtype=object)

In [35]:
base_total["tipo_eleccion"].unique()

array(['PRESIDENCIAL', 'PARLAMENTARIA', 'MUNICIPALES', 'CORE'],
      dtype=object)

In [36]:
print(base_total[base_total.tipo_voto_presidencial.notnull()]["tipo_eleccion"].unique())
print(base_total[base_total.tipo_voto_presidencial.isnull()]["tipo_eleccion"].unique())

['PRESIDENCIAL']
['PARLAMENTARIA' 'MUNICIPALES' 'CORE']


In [37]:
print(base_total[base_total.tipo_voto_presidencial.notnull()].shape)
print(base_total[base_total.tipo_voto_presidencial.isnull()].shape)

(65906, 17)
(308814, 17)


In [38]:
tmp_1 = base_total[base_total.tipo_voto_presidencial.notnull()].shape[0]
tmp_2 = base_total[base_total.tipo_voto_presidencial.isnull()].shape[0]

total_records == tmp_1 + tmp_2

True

In [39]:
del tmp_1, tmp_2

Para poder obtener mejores conclusiones respecto de eso, voy a convertir la columna `base_total`.`fecha` en DateTime de Pandas.

In [40]:
base_total["date"] = pd.to_datetime(base_total["fecha"], format = "%d/%m/%Y")

In [41]:
base_total["date"].unique()

array(['1989-12-14T00:00:00.000000000', '1993-12-11T00:00:00.000000000',
       '1999-12-12T00:00:00.000000000', '2000-01-16T00:00:00.000000000',
       '2005-12-11T00:00:00.000000000', '2006-01-15T00:00:00.000000000',
       '2009-12-13T00:00:00.000000000', '2010-01-17T00:00:00.000000000',
       '2013-11-17T00:00:00.000000000', '2013-12-15T00:00:00.000000000',
       '2017-11-19T00:00:00.000000000', '2017-12-17T00:00:00.000000000',
       '1997-12-11T00:00:00.000000000', '2001-12-16T00:00:00.000000000',
       '2004-10-31T00:00:00.000000000', '2008-10-26T00:00:00.000000000',
       '2012-10-28T00:00:00.000000000', '2016-10-23T00:00:00.000000000',
       '1992-10-28T00:00:00.000000000', '1996-10-27T00:00:00.000000000',
       '2000-10-29T00:00:00.000000000'], dtype='datetime64[ns]')

In [42]:
base_total.drop(columns = ["fecha"], inplace = True)

Verificamos la cantidad de periodos y de fechas de votación (`periodo`, `date`).

In [43]:
base_total["periodo"].unique()

array(['1990 a 1994', '1994 a 2000', '2000 a 2006', '2006 a 2010',
       '2010 a 2014', '2014 a 2018', '2018-2022', '2018 a 2022',
       '1990-1994', '1994-2002', '1998-2006', '2002-2010', '2006-2014',
       '2010-2018', '2014-2022', '2018-2020', '1994-1998', '1998-2002',
       '2002-2006', '2006-2010', '2010-2014', '2014-2018', '2004-2008',
       '2008-2012', '2012-2016', '2016-2020', '1992-1996', '1996-2000',
       '2000-2004'], dtype=object)

In [44]:
base_total.groupby(["tipo_eleccion", "periodo", "date"])["candidato"].count()

tipo_eleccion  periodo      date      
CORE           2014-2018    2013-11-17    14578
               2018-2022    2017-11-19    19192
MUNICIPALES    1992-1996    1992-10-28     7164
               1996-2000    1996-10-27    18542
               2000-2004    2000-10-29    15766
               2004-2008    2004-10-31    27314
               2008-2012    2008-10-26    40013
               2012-2016    2012-10-28    20942
               2016-2020    2016-10-23    29377
PARLAMENTARIA  1990-1994    1989-12-14    15465
               1994-1998    1993-12-11     7903
               1994-2002    1993-12-11     3049
               1998-2002    1997-12-11     7203
               1998-2006    1997-12-11     5636
               2002-2006    2001-12-16     8224
               2002-2010    2001-12-16     2740
               2006-2010    2005-12-11     9327
               2006-2014    2005-12-11     5913
               2010-2014    2009-12-13     9901
               2010-2018    2009-12-13     3382
 

Voy a cambiar algo de los strings, para que la columna `periodo` tenga valores consistentes.

In [45]:
base_total["periodo"] = base_total["periodo"].str.replace(" a ", "-")

In [46]:
base_total.groupby(["tipo_eleccion", "periodo", "date"])["candidato"].count()

tipo_eleccion  periodo    date      
CORE           2014-2018  2013-11-17    14578
               2018-2022  2017-11-19    19192
MUNICIPALES    1992-1996  1992-10-28     7164
               1996-2000  1996-10-27    18542
               2000-2004  2000-10-29    15766
               2004-2008  2004-10-31    27314
               2008-2012  2008-10-26    40013
               2012-2016  2012-10-28    20942
               2016-2020  2016-10-23    29377
PARLAMENTARIA  1990-1994  1989-12-14    15465
               1994-1998  1993-12-11     7903
               1994-2002  1993-12-11     3049
               1998-2002  1997-12-11     7203
               1998-2006  1997-12-11     5636
               2002-2006  2001-12-16     8224
               2002-2010  2001-12-16     2740
               2006-2010  2005-12-11     9327
               2006-2014  2005-12-11     5913
               2010-2014  2009-12-13     9901
               2010-2018  2009-12-13     3382
               2014-2018  2013-11-17     53

In [47]:
base_total.groupby(["tipo_eleccion", "periodo", "date"])["votos"].sum()

tipo_eleccion  periodo    date      
CORE           2014-2018  2013-11-17     6685458
               2018-2022  2017-11-19     6674828
MUNICIPALES    1992-1996  1992-10-28     7040753
               1996-2000  1996-10-27     7092826
               2000-2004  2000-10-29     7089775
               2004-2008  2004-10-31    13746062
               2008-2012  2008-10-26    13909503
               2012-2016  2012-10-28    11561404
               2016-2020  2016-10-23     9842471
PARLAMENTARIA  1990-1994  1989-12-14    14316682
               1994-1998  1993-12-11     7385016
               1994-2002  1993-12-11     2045681
               1998-2002  1997-12-11     7046361
               1998-2006  1997-12-11     5102906
               2002-2006  2001-12-16     7034292
               2002-2010  2001-12-16     1975017
               2006-2010  2005-12-11     7207351
               2006-2014  2005-12-11     5182224
               2010-2014  2009-12-13     7263537
               2010-2018  2009-1

In [48]:
pd.pivot_table(data = base_total, 
               values = ["votos"], 
               index = ["tipo_eleccion", "periodo", "date"], 
               aggfunc = np.sum, 
               margins = True)

votos
tipo_eleccion periodo   date                 
CORE          2014-2018 2013-11-17    6685458
              2018-2022 2017-11-19    6674828
MUNICIPALES   1992-1996 1992-10-28    7040753
              1996-2000 1996-10-27    7092826
              2000-2004 2000-10-29    7089775
              2004-2008 2004-10-31   13746062
              2008-2012 2008-10-26   13909503
              2012-2016 2012-10-28   11561404
              2016-2020 2016-10-23    9842471
PARLAMENTARIA 1990-1994 1989-12-14   14316682
              1994-1998 1993-12-11    7385016
              1994-2002 1993-12-11    2045681
              1998-2002 1997-12-11    7046361
              1998-2006 1997-12-11    5102906
              2002-2006 2001-12-16    7034292
              2002-2010 2001-12-16    1975017
              2006-2010 2005-12-11    7207351
              2006-2014 2005-12-11    5182224
              2010-2014 2009-12-13    7263537
              2010-2018 2009-12-13    2053480
              2014-2018 2013-11-17    6698524
              2014-2022 2013-11-17    4852165
              2018-2020 2017-11-19    8492876
PRESIDENCIAL  1990-1994 1989-12-14    7158727
              1994-2000 1993-12-11    7383286
              2000-2006 1999-12-12    7271572
                        2000-01-16    7326753
              2006-2010 2005-12-11    7207278
                        2006-01-15    7162345
              2010-2014 2009-12-13    7264136
                        2010-01-17    7203371
              2014-2018 2013-11-17    6699011
                        2013-12-15    5697751
              2018-2022 2017-11-19    6703327
                        2017-12-17    7032878
All                     NaT         254409627

Con las columnas `tipo_eleccion` y `periodo`, se puede validar que la votación corresponde a una primera o una segunda vuelta. Ahora, nos movemos a limpiar los datos de `partido`.

In [49]:
base_total["partido"].unique()

array(['INDEPENDIENTE', 'DEMOCRATA CRISTIANO', nan,
       'ALIANZA HUMANISTA VERDE', 'UNION DE CENTRO CENTRO',
       'COMUNISTA DE CHILE', 'UNION DEMOCRATA INDEPENDIENTE',
       'POR LA DEMOCRACIA', 'HUMANISTA', 'UNION DEMOCRATA INDEPENDIENTE ',
       'POR LA DEMOCRACIA ', 'SOCIALISTA DE CHILE', 'RENOVACION NACIONAL',
       'PARTIDO ECOLOGISTA VERDE', 'PARTIDO HUMANISTA',
       'PARTIDO PROGRESISTA', 'PARTIDO SOCIALISTA DE CHILE',
       'PARTIDO REGIONALISTA DE LOS INDEPENDIENTES', 'PARTIDO IGUALDAD',
       'INDEPENDIENTES', 'PAIS', 'PARTIDO DEMOCRATA CRISTIANO',
       'UNION PATRIOTICA', '  ', 'RADICAL', 'P.A.I.S.', 'LIBERAL',
       'DEMOCRACIA RADICAL', 'NACIONAL', 'SOCIALISTA CHILENO', 'DEL SUR',
       'AVANZADA NACIONAL', 'ALIANZA HUMANISTA-VERDE', ' ',
       'RADICAL DE CHILE', 'MOVIMIENTO DE ACCION POPULAR U',
       'UNION DE CENTRO CENTRO PROGRES', 'RADICAL SOCIALDEMOCRATA',
       'ALIANZA NACIONAL DE LOS INDEPE', 'REGIONALISTA DE LOS INDEPENDIE',
       'PARTIDO P

In [50]:
base_total.groupby(["tipo_eleccion", "partido"])["candidato"].count()

tipo_eleccion  partido                       
CORE           AMPLITUD                           383
               CANDIDATURA INDEPENDIENTE           68
               CIUDADANOS                         116
               DEMOCRACIA REGIONAL PATAGONICA     233
               ECOLOGISTA VERDE DEL NORTE          22
                                                 ... 
PRESIDENCIAL   SOCIALISTA DE CHILE               2211
               UNION DE CENTRO CENTRO            1016
               UNION DEMOCRATA INDEPENDIENTE     3265
               UNION DEMOCRATA INDEPENDIENTE     1018
               UNION PATRIOTICA                   736
Name: candidato, Length: 213, dtype: int64

Hay algunos nombres de partidos sospechosos, como `TODOS`, ` ` (un espacio) y `  ` (dos espacios). Voy a explorar esos valores.

In [51]:
base_total.query("partido == 'TODOS' or partido == ' ' or partido == '  '"). \
           groupby(["tipo_eleccion", "date", "partido", "candidato", "cargo"])["votos"].sum()

tipo_eleccion  date        partido  candidato                 cargo   
MUNICIPALES    1992-10-28           VOTOS EN BLANCO           CONCEJAL    214032
                                    VOTOS NULOS               CONCEJAL    418775
               1996-10-27           VOTOS EN BLANCO           CONCEJAL    217027
                                    VOTOS NULOS               CONCEJAL    562400
               2000-10-29           VOTOS EN BLANCO           CONCEJAL    181985
                                    VOTOS NULOS               CONCEJAL    392319
               2004-10-31           VOTOS EN BLANCO           ALCALDE     187672
                                                              CONCEJAL    272966
                                    VOTOS NULOS               ALCALDE     374797
                                                              CONCEJAL    477847
               2008-10-26           VOTOS EN BLANCO           ALCALDE     189770
                                      

El `TODOS` es un partido, que desconocia.

Existen unas entradas identificadas como `INDEPENDIENTES`, que voy a explorar ahora.

In [52]:
base_total.query("partido == 'INDEPENDIENTES'")["tipo_eleccion"].unique()

array(['PRESIDENCIAL', 'CORE'], dtype=object)

In [53]:
base_total.query("partido == 'INDEPENDIENTES'")["date"].unique()

array(['2017-11-19T00:00:00.000000000', '2017-12-17T00:00:00.000000000',
       '2013-11-17T00:00:00.000000000'], dtype='datetime64[ns]')

In [54]:
base_total.query("partido == 'INDEPENDIENTES' and tipo_eleccion == 'PRESIDENCIAL'"). \
           groupby(["partido", "candidato", "periodo"])["candidato"].count()

partido         candidato                    periodo  
INDEPENDIENTES  ALEJANDRO  GUILLIER ALVAREZ  2018-2022    1472
                BEATRIZ SANCHEZ MUÑOZ        2018-2022     736
                JOSE ANTONIO KAST RIST       2018-2022     736
                SEBASTIAN PIÑERA ECHENIQUE   2018-2022    1472
Name: candidato, dtype: int64

In [55]:
base_total.query("periodo == '2018-2022' and tipo_eleccion == 'PRESIDENCIAL'"). \
           groupby(["partido", "candidato"])["candidato"].count()

partido                      candidato                      
INDEPENDIENTES               ALEJANDRO  GUILLIER ALVAREZ        1472
                             BEATRIZ SANCHEZ MUÑOZ               736
                             JOSE ANTONIO KAST RIST              736
                             SEBASTIAN PIÑERA ECHENIQUE         1472
PAIS                         ALEJANDRO NAVARRO BRAIN             736
PARTIDO DEMOCRATA CRISTIANO  CAROLINA GOIC BOROEVIC              736
PARTIDO PROGRESISTA          MARCO  ENRIQUEZ-OMINAMI GUMUCIO     736
UNION PATRIOTICA             EDUARDO ARTES BRICHETTI             736
Name: candidato, dtype: int64

Los candidatos para el 2018 Alejandro Guillier, Beatriz Sánchez, Jose Antonio Kast y Sebastian Piñera eran independientes, que es algo que desconocía.

Voy a revisar ahora las diferencias que hayan entre el `PARTIDO LIBERAL` y el `PARTIDO LIBERAL DE CHILE`.

In [56]:
base_total.query("partido == 'LIBERAL' or partido == 'LIBERAL DE CHILE'"). \
           groupby(["tipo_eleccion", "partido", "candidato", "periodo"])["candidato"].count()

tipo_eleccion  partido           candidato                       periodo  
MUNICIPALES    LIBERAL           AGUILAR RIQUELME DANIEL         1992-1996     1
                                 ARAVENA SAN MARTIN JUAN CARLOS  1992-1996     1
                                 BARRIGA SOTO TERESA             1992-1996     1
                                 BOGDANIC BASSI LUIS             1992-1996     1
                                 BORGNA MOLINA LUIS              1992-1996     1
                                                                              ..
PARLAMENTARIA  LIBERAL DE CHILE  MIROSEVIC VERDUGO VLADO         2018-2020     9
                                 MORAN MORAN IVAN                2018-2020    35
                                 MUÑOZ RETAMALES EDGAR           2018-2020    15
                                 NAVARRO AGUAYO CAMILA           2018-2020     9
                                 VIGUERAS CHERRES ROXANA         2018-2020     9
Name: candidato, Length: 68, dtype

In [57]:
pd.pivot_table(data = base_total.query("partido == 'LIBERAL' or partido == 'LIBERAL DE CHILE'"),
               values = "votos",
               index = ["tipo_eleccion", "periodo"],
               columns = "partido",
               aggfunc = len)

partido                  LIBERAL  LIBERAL DE CHILE
tipo_eleccion periodo                             
MUNICIPALES   1992-1996     35.0               NaN
              2000-2004     12.0               NaN
PARLAMENTARIA 1990-1994    329.0               NaN
              2002-2006     50.0               NaN
              2002-2010     32.0               NaN
              2018-2020      NaN             137.0

Pareciera ser que son dos partidos diferentes, así que los etiquetaré de manera independiente. Finalmente, revisaré los votos que aparecen con partido `nan`.

In [58]:
pd.pivot_table(data = base_total.query("partido.isnull()"),
               values = "votos",
               index = ["tipo_eleccion", "periodo"],
               columns = ["candidato"],
               aggfunc = len)

candidato                BRANDON JOHAN  CERDA ALVARADO  CARLOS  LOPEZ VEGA  \
tipo_eleccion periodo                                                        
CORE          2014-2018                            NaN                 NaN   
              2018-2022                            NaN                 NaN   
MUNICIPALES   2012-2016                            NaN                 NaN   
              2016-2020                            5.0                 2.0   
PARLAMENTARIA 2002-2006                            NaN                 NaN   
              2002-2010                            NaN                 NaN   
              2018-2020                            NaN                 NaN   
PRESIDENCIAL  1990-1994                            NaN                 NaN   
              1994-2000                            NaN                 NaN   
              2000-2006                            NaN                 NaN   
              2006-2010                            NaN                 NaN   
              2010-2014                            NaN                 NaN   
              2014-2018                            NaN                 NaN   
              2018-2022                            NaN                 NaN   

candidato                LUIS  MOYANO CRUZ  MIREYA  CORTES MILLA  \
tipo_eleccion periodo                                              
CORE          2014-2018                NaN                   NaN   
              2018-2022                NaN                   NaN   
MUNICIPALES   2012-2016                NaN                   NaN   
              2016-2020                1.0                   1.0   
PARLAMENTARIA 2002-2006                NaN                   NaN   
              2002-2010                NaN                   NaN   
              2018-2020                NaN                   NaN   
PRESIDENCIAL  1990-1994                NaN                   NaN   
              1994-2000                NaN                   NaN   
              2000-2006                NaN                   NaN   
              2006-2010                NaN                   NaN   
              2010-2014                NaN                   NaN   
              2014-2018                NaN                   NaN   
              2018-2022                NaN                   NaN   

candidato                OSVALDO  VILLACORTA TORO  VOTOS EN BLANCO  \
tipo_eleccion periodo                                                
CORE          2014-2018                       NaN            573.0   
              2018-2022                       NaN            626.0   
MUNICIPALES   2012-2016                       NaN           1134.0   
              2016-2020                       1.0           1218.0   
PARLAMENTARIA 2002-2006                       NaN           1026.0   
              2002-2010                       NaN            386.0   
              2018-2020                       NaN            854.0   
PRESIDENCIAL  1990-1994                       NaN              NaN   
              1994-2000                       NaN              NaN   
              2000-2006                       NaN              NaN   
              2006-2010                       NaN              NaN   
              2010-2014                       NaN              NaN   
              2014-2018                       NaN           1144.0   
              2018-2022                       NaN           1472.0   

candidato                VOTOS EN BLANCO   VOTOS NULOS  VOTOS NULOS   
tipo_eleccion periodo                                                 
CORE          2014-2018               NaN        573.0           NaN  
              2018-2022               NaN        626.0           NaN  
MUNICIPALES   2012-2016               NaN       1134.0           NaN  
              2016-2020               NaN       1218.0           NaN  
PARLAMENTARIA 2002-2006               NaN       1026.0           NaN  
              2002-2010               NaN   

In [59]:
pd.pivot_table(data = base_total.query("partido.isnull()"),
               values = "votos",
               index = ["tipo_eleccion", "cargo", "periodo"],
               columns = ["candidato", "comuna"],
               aggfunc = len)

candidato                                  BRANDON JOHAN  CERDA ALVARADO  \
comuna                                                           CALBUCO   
tipo_eleccion cargo              periodo                                   
CORE          CONSEJERO REGIONAL 2014-2018                           NaN   
                                 2018-2022                           NaN   
MUNICIPALES   ALCALDE            2012-2016                           NaN   
                                 2016-2020                           NaN   
              CONCEJAL           2012-2016                           NaN   
                                 2016-2020                           5.0   
PARLAMENTARIA DIPUTADO           2002-2006                           NaN   
                                 2018-2020                           NaN   
              SENADOR            2002-2010                           NaN   
                                 2018-2020                           NaN   
PRESIDENCIAL  PRESIDENTE         1990-1994                           NaN   
                                 1994-2000                           NaN   
                                 2000-2006                           NaN   
                                 2006-2010                           NaN   
                                 2010-2014                           NaN   
                                 2014-2018                           NaN   
                                 2018-2022                           NaN   

candidato                                  CARLOS  LOPEZ VEGA  \
comuna                                           SIERRA GORDA   
tipo_eleccion cargo              periodo                        
CORE          CONSEJERO REGIONAL 2014-2018                NaN   
                                 2018-2022                NaN   
MUNICIPALES   ALCALDE            2012-2016                NaN   
                                 2016-2020                2.0   
              CONCEJAL           2012-2016                NaN   
                                 2016-2020                NaN   
PARLAMENTARIA DIPUTADO           2002-2006                NaN   
                                 2018-2020                NaN   
              SENADOR            2002-2010                NaN   
                                 2018-2020                NaN   
PRESIDENCIAL  PRESIDENTE         1990-1994                NaN   
                                 1994-2000                NaN   
                                 2000-2006                NaN   
                                 2006-2010                NaN   
                                 2010-2014                NaN   
                                 2014-2018                NaN   
                                 2018-2022                NaN   

candidato                                  LUIS  MOYANO CRUZ  \
comuna                                             TOCOPILLA   
tipo_eleccion cargo              periodo                       
CORE          CONSEJERO REGIONAL 2014-2018               NaN   
                                 2018-2022               NaN   
MUNICIPALES   ALCALDE            2012-2016               NaN   
                                 2016-2020               1.0   
              CONCEJAL           2012-2016               NaN   
                                 2016-2020               NaN   
PARLAMENTARIA DIPUTADO           2002-2006               NaN   
                                 2018-2020               NaN   
              SENADOR            2002-2010               NaN   
                                 2018-2020               NaN   
PRESIDENCIAL  PRESIDENTE         1990-1994               NaN   
                                 1994-2000               NaN   
                                 2000-2006               NaN   
                                 2006-2010               NaN   
                                 2010-2014               NaN   
                          

* `BRANDON JOHAN CERDA ALVARADO`, en las municipales de 2016, fue candidato por el Partido Comunista de Chile, según TVN: https://www.24horas.cl/municipales-2016/resultados/los-lagos/article2169202.ece
* `CARLOS LOPEZ VEGA`, en las municipales de 2016, fue candidato independiente, según TVN: https://www.24horas.cl/municipales-2016/resultados/antofagasta/article2169322.ece. También verifican la independencia SoyChile: https://www.soychile.cl/Antofagasta/Politica/2016/10/24/425718/Carlos-Lopez-aseguro-que-gano-en-Sierra-Gorda-y-pedira-recuento-de-votos-al-Tribunal-Electoral.aspx
* `LUIS MOYANO CRUZ`, en las municipales de 2016, fue candidato independiente, según TVN: https://www.24horas.cl/municipales-2016/resultados/antofagasta/article2169315.ece. Wikipedia también da cuenta de su independencia: https://es.m.wikipedia.org/wiki/Elecciones_municipales_de_2016_en_la_Regi%C3%B3n_de_Antofagasta
* `MIREYA CORTES MILLA`, en las municipales de 2016, fue candidata por el Partido Progresista, según TVN: https://www.24horas.cl/municipales-2016/resultados/coquimbo/article2169337.ece
* `OSVALDO VILLACORTA TORO`, en las municipales de 2016, fue candidato independiente, según TVN: https://www.24horas.cl/municipales-2016/resultados/antofagasta/article2169315.ece

Dichos candidatos, para la combinación específica que aparece en los datos, deben ver su `partido` modificado de manera acorde.

In [60]:
base_total[(base_total["candidato"].str.contains("BRANDON JOHAN")) & \
           (base_total.comuna == "CALBUCO") & \
           (base_total.periodo == "2016-2020")]

index                      candidato     cargo   comuna  distrito  \
273990  73452  BRANDON JOHAN  CERDA ALVARADO  CONCEJAL  CALBUCO      26.0   
274044  73506  BRANDON JOHAN  CERDA ALVARADO  CONCEJAL  CALBUCO      26.0   
274098  73560  BRANDON JOHAN  CERDA ALVARADO  CONCEJAL  CALBUCO      26.0   
274152  73614  BRANDON JOHAN  CERDA ALVARADO  CONCEJAL  CALBUCO      26.0   
274206  73668  BRANDON JOHAN  CERDA ALVARADO  CONCEJAL  CALBUCO      26.0   

       lista circunscripcion provincia partido    periodo        region sexo  \
273990     S         CALBUCO       NaN     NaN  2016-2020  DE LOS LAGOS  NaN   
274044     S          COLACO       NaN     NaN  2016-2020  DE LOS LAGOS  NaN   
274098     S       ISLA GUAR       NaN     NaN  2016-2020  DE LOS LAGOS  NaN   
274152     S    ISLA PULUQUI       NaN     NaN  2016-2020  DE LOS LAGOS  NaN   
274206     S          PARGUA       NaN     NaN  2016-2020  DE LOS LAGOS  NaN   

       tipo_eleccion tipo_voto_presidencial  votos electo       date  
273990   MUNICIPALES                    NaN     83     si 2016-10-23  
274044   MUNICIPALES                    NaN      0     si 2016-10-23  
274098   MUNICIPALES                    NaN      1     si 2016-10-23  
274152   MUNICIPALES                    NaN      1     si 2016-10-23  
274206   MUNICIPALES                    NaN      1     si 2016-10-23

In [61]:
base_total[(base_total["candidato"].str.contains("LOPEZ VEGA")) & \
           (base_total.comuna == "SIERRA GORDA") & \
           (base_total.periodo == "2016-2020")]

index           candidato    cargo        comuna  distrito lista  \
197316  15484  CARLOS  LOPEZ VEGA  ALCALDE  SIERRA GORDA       3.0     T   
197321  15489  CARLOS  LOPEZ VEGA  ALCALDE  SIERRA GORDA       3.0     T   

           circunscripcion provincia partido    periodo          region sexo  \
197316        SIERRA GORDA       NaN     NaN  2016-2020  DE ANTOFAGASTA  NaN   
197321  SIERRA GORDA NORTE       NaN     NaN  2016-2020  DE ANTOFAGASTA  NaN   

       tipo_eleccion tipo_voto_presidencial  votos electo       date  
197316   MUNICIPALES                    NaN    443     no 2016-10-23  
197321   MUNICIPALES                    NaN    181     no 2016-10-23

In [62]:
base_total[(base_total["candidato"].str.contains("MOYANO CRUZ")) & \
           (base_total.comuna == "TOCOPILLA") & \
           (base_total.periodo == "2016-2020")]

index          candidato    cargo     comuna  distrito lista  \
197331  15499  LUIS  MOYANO CRUZ  ALCALDE  TOCOPILLA       3.0     T   

       circunscripcion provincia partido    periodo          region sexo  \
197331       TOCOPILLA       NaN     NaN  2016-2020  DE ANTOFAGASTA  NaN   

       tipo_eleccion tipo_voto_presidencial  votos electo       date  
197331   MUNICIPALES                    NaN   3621     si 2016-10-23

In [63]:
base_total[(base_total["candidato"].str.contains("CORTES MILLA")) & \
           (base_total.comuna == "VICUÑA") & \
           (base_total.periodo == "2016-2020")]

index             candidato     cargo  comuna  distrito lista  \
287302  86764  MIREYA  CORTES MILLA  CONCEJAL  VICUÑA       5.0     O   

       circunscripcion provincia partido    periodo       region sexo  \
287302          VICUÑA       NaN     NaN  2016-2020  DE COQUIMBO  NaN   

       tipo_eleccion tipo_voto_presidencial  votos electo       date  
287302   MUNICIPALES                    NaN     39     si 2016-10-23

In [64]:
base_total[(base_total["candidato"].str.contains("VILLACORTA TORO")) & \
           (base_total.comuna == "TOCOPILLA") & \
           (base_total.periodo == "2016-2020")]

index                 candidato    cargo     comuna  distrito lista  \
197335  15503  OSVALDO  VILLACORTA TORO  ALCALDE  TOCOPILLA       3.0     T   

       circunscripcion provincia partido    periodo          region sexo  \
197335       TOCOPILLA       NaN     NaN  2016-2020  DE ANTOFAGASTA  NaN   

       tipo_eleccion tipo_voto_presidencial  votos electo       date  
197335   MUNICIPALES                    NaN     95     no 2016-10-23

In [65]:
base_total[base_total["partido"].isnull()].groupby("candidato").groups.keys()

dict_keys(['BRANDON JOHAN  CERDA ALVARADO', 'CARLOS  LOPEZ VEGA', 'LUIS  MOYANO CRUZ', 'MIREYA  CORTES MILLA', 'OSVALDO  VILLACORTA TORO', 'VOTOS EN BLANCO', 'VOTOS EN BLANCO ', 'VOTOS NULOS', 'VOTOS NULOS '])

In [66]:
condiciones = (base_total["candidato"].str.contains("BRANDON JOHAN")) & \
           (base_total.comuna == "CALBUCO") & \
           (base_total.periodo == "2016-2020")

print(base_total.loc[condiciones, "partido"])

base_total.loc[condiciones, "partido"] = "COMUNISTA DE CHILE"

print(base_total.loc[condiciones, "partido"])

273990    NaN
274044    NaN
274098    NaN
274152    NaN
274206    NaN
Name: partido, dtype: object
273990    COMUNISTA DE CHILE
274044    COMUNISTA DE CHILE
274098    COMUNISTA DE CHILE
274152    COMUNISTA DE CHILE
274206    COMUNISTA DE CHILE
Name: partido, dtype: object


In [67]:
condiciones = (base_total["candidato"].str.contains("LOPEZ VEGA")) & \
           (base_total.comuna == "SIERRA GORDA") & \
           (base_total.periodo == "2016-2020")

print(base_total.loc[condiciones, "partido"])

base_total.loc[condiciones, "partido"] = "INDEPENDIENTE"

print(base_total.loc[condiciones, "partido"])

197316    NaN
197321    NaN
Name: partido, dtype: object
197316    INDEPENDIENTE
197321    INDEPENDIENTE
Name: partido, dtype: object


In [68]:
condiciones = (base_total["candidato"].str.contains("MOYANO CRUZ")) & \
           (base_total.comuna == "TOCOPILLA") & \
           (base_total.periodo == "2016-2020")

print(base_total.loc[condiciones, "partido"])

base_total.loc[condiciones, "partido"] = "INDEPENDIENTE"

print(base_total.loc[condiciones, "partido"])

197331    NaN
Name: partido, dtype: object
197331    INDEPENDIENTE
Name: partido, dtype: object


In [69]:
condiciones = (base_total["candidato"].str.contains("CORTES MILLA")) & \
           (base_total.comuna == "VICUÑA") & \
           (base_total.periodo == "2016-2020")

print(base_total.loc[condiciones, "partido"])

base_total.loc[condiciones, "partido"] = "PROGRESISTA"

print(base_total.loc[condiciones, "partido"])

287302    NaN
Name: partido, dtype: object
287302    PROGRESISTA
Name: partido, dtype: object


In [70]:
condiciones = (base_total["candidato"].str.contains("VILLACORTA TORO")) & \
           (base_total.comuna == "TOCOPILLA") & \
           (base_total.periodo == "2016-2020")

print(base_total.loc[condiciones, "partido"])

base_total.loc[condiciones, "partido"] = "INDEPENDIENTE"

print(base_total.loc[condiciones, "partido"])

197335    NaN
Name: partido, dtype: object
197335    INDEPENDIENTE
Name: partido, dtype: object


In [71]:
base_total[base_total["partido"].isnull()].groupby("candidato").groups.keys()

dict_keys(['VOTOS EN BLANCO', 'VOTOS EN BLANCO ', 'VOTOS NULOS', 'VOTOS NULOS '])

Verificamos los valores de `candidato` para votos nulos y blancos.

In [72]:
base_total[base_total["candidato"].str.contains("VOTO")].groupby(["candidato", "partido"]).groups.keys()

dict_keys([('VOTOS NULOS ', nan), ('VOTOS EN BLANCO ', nan), ('VOTOS NULOS', nan), ('VOTOS EN BLANCO', nan), ('VOTOS NULOS', '  '), ('VOTOS EN BLANCO', '  '), ('VOTOS EN BLANCO', ' '), ('VOTOS NULOS', ' ')])

In [73]:
base_total.loc[(base_total["candidato"].str.contains("VOTO")) & \
               (base_total.partido.isnull()), "partido"] = " "

In [74]:
base_total[base_total["candidato"].str.contains("VOTO")].groupby(["candidato", "partido"]).groups.keys()

dict_keys([('VOTOS EN BLANCO', ' '), ('VOTOS EN BLANCO', '  '), ('VOTOS EN BLANCO ', ' '), ('VOTOS NULOS', ' '), ('VOTOS NULOS', '  '), ('VOTOS NULOS ', ' ')])

Hice un mapeo de los partidos, para asignar etiquetas únicas. Este mapeo estará en un diccionario.

In [75]:
mapeo_partidos = {
    ' ': 'NULOS/BLANCOS',
    '  ': 'NULOS/BLANCOS',
    'ALIANZA HUMANISTA VERDE': 'ALIANZA HUMANISTA VERDE',
    'ALIANZA HUMANISTA-VERDE': 'ALIANZA HUMANISTA VERDE',
    'ALIANZA NACIONAL DE LOS INDEPE': 'ALIANZA NACIONAL DE LOS INDEPENDIENTES',
    'ALIANZA NACIONAL DE LOS INDEPENDIENTES': 'ALIANZA NACIONAL DE LOS INDEPENDIENTES',
    'AMPLITUD': 'AMPLITUD',
    'AVANZADA NACIONAL': 'AVANZADA NACIONAL',
    'CANDIDATURA INDEPENDIENTE': 'INDEPENDIENTE',
    'CHILEPRIMERO': 'CHILEPRIMERO',
    'CIUDADANOS': 'CIUDADANOS',
    'COMUNISTA DE CHILE': 'PARTIDO COMUNISTA DE CHILE',
    'DE ACCION REGIONALISTA DE CHIL': 'PARTIDO DE ACCION REGIONALISTA DE CHILE',
    'DE ACCION REGIONALISTA DE CHILE': 'PARTIDO DE ACCION REGIONALISTA DE CHILE',
    'DE TRABAJADORES REVOLUCIONARIOS': 'PARTIDO DE TRABAJADORES REVOLUCIONARIOS',
    'DEL SUR': 'PARTIDO DEL SUR',
    'DEMOCRACIA RADICAL': 'DEMOCRACIA RADICAL',
    'DEMOCRACIA REGIONAL PATAGONICA': 'DEMOCRACIA REGIONAL PATAGONICA',
    'DEMOCRACIA REGIONAL PATAGONICA ': 'DEMOCRACIA REGIONAL PATAGONICA',
    'DEMOCRATA CRISTIANO': 'DEMOCRACIA CRISTIANA',
    'ECOLOGISTA': 'PARTIDO ECOLOGISTA',
    'ECOLOGISTA VERDE': 'PARTIDO ECOLOGISTA VERDE',
    'ECOLOGISTA VERDE DEL NORTE': 'PARTIDO ECOLOGISTA VERDE DEL NORTE',
    'EVOLUCION POLITICA': 'EVOLUCION POLITICA',
    'FEDERACION REGIONALISTA VERDE SOCIAL': 'FEDERACION REGIONALISTA VERDE SOCIAL',
    'FRENTE REGIONAL Y POPULAR': 'FRENTE REGIONAL Y POPULAR',
    'FUERZA DEL NORTE': 'FUERZA DEL NORTE',
    'FUERZA PAIS': 'FUERZA PAIS',
    'FUERZA REGIONAL NORTE VERDE': 'FUERZA REGIONAL NORTE VERDE',
    'HUMANISTA': 'PARTIDO HUMANISTA',
    'IGUALDAD': 'PARTIDO IGUALDAD',
    'INDEPENDIENTE': 'INDEPENDIENTE',
    'INDEPENDIENTE AMPLITUD': 'INDEPENDIENTE - AMPLITUR',
    'INDEPENDIENTE CIUDADANOS': 'INDEPENDIENTE - CIUDADANOS',
    'INDEPENDIENTE COMUNISTA DE CHILE': 'INDEPENDIENTE - PARTIDO COMUNISTA DE CHILE',
    'INDEPENDIENTE DEMOCRACIA REGIONAL PATAGONICA': 'INDEPENDIENTE - DEMOCRACIA REGIONAL PATAGONICA',
    'INDEPENDIENTE DEMOCRATA CRISTIANO': 'INDEPENDIENTE - DEMOCRACIA CRISTIANA',
    'INDEPENDIENTE ECOLOGISTA VERDE': 'INDEPENDIENTE - PARTIDO ECOLOGISTA VERDE',
    'INDEPENDIENTE EVOLUCION POLITICA': 'INDEPENDIENTE - EVOLUCION POLITICA',
    'INDEPENDIENTE FEDERACION REGIONALISTA VERDE SOCIAL': 'INDEPENDIENTE - FEDERACION REGIONALISTA VERDE SOCIAL',
    'INDEPENDIENTE HUMANISTA': 'INDEPENDIENTE - PARTIDO HUMANISTA',
    'INDEPENDIENTE IGUALDAD': 'INDEPENDIENTE - PARTIDO IGUALDAD',
    'INDEPENDIENTE IZQUIERDA CIUDADANA DE CHILE': 'INDEPENDIENTE - IZQUIERDA CIUDADANA DE CHILE',
    'INDEPENDIENTE LIBERAL DE CHILE': 'INDEPENDIENTE - PARTIDO LIBERAL DE CHILE',
    'INDEPENDIENTE MAS REGION': 'INDEPENDIENTE - MAS REGION',
    'INDEPENDIENTE PAIS': 'INDEPENDIENTE - PARTIDO PAIS',
    'INDEPENDIENTE PODER': 'INDEPENDIENTE - PODER',
    'INDEPENDIENTE POR LA DEMOCRACIA': 'INDEPENDIENTE - PPD',
    'INDEPENDIENTE PROGRESISTA': 'INDEPENDIENTE - PARTIDO PROGRESISTA',
    'INDEPENDIENTE RADICAL SOCIALDEMOCRATA': 'INDEPENDIENTE - PARTIDO RADICAL SOCIALDEMOCRATA',
    'INDEPENDIENTE REGIONALISTA INDEPENDIENTE': 'INDEPENDIENTE - PARTIDO REGIONALISTA INDEPENDIENTE',
    'INDEPENDIENTE RENOVACION NACIONAL': 'INDEPENDIENTE - RENOVACION NACIONAL',
    'INDEPENDIENTE REVOLUCION DEMOCRATICA': 'INDEPENDIENTE - REVOLUCION DEMOCRATICA',
    'INDEPENDIENTE SOCIALISTA DE CHILE': 'INDEPENDIENTE - PARTIDO SOCIALISTA',
    'INDEPENDIENTE TODOS': 'INDEPENDIENTE - TODOS',
    'INDEPENDIENTE UNION DEMOCRATA INDEPENDIENTE': 'INDEPENDIENTE - UDI',
    'INDEPENDIENTES': 'INDEPENDIENTE',
    'INDEPENDIENTES LISTA A': 'INDEPENDIENTE - LISTA A',
    'INDEPENDIENTES LISTA B': 'INDEPENDIENTE - LISTA B',
    'INDEPENDIENTES LISTA D': 'INDEPENDIENTE - LISTA D',
    'INDEPENDIENTES LISTA G': 'INDEPENDIENTE - LISTA G',
    'INDEPENDIENTES LISTA H': 'INDEPENDIENTE - LISTA H',
    'INDEPENDIENTES LISTA I': 'INDEPENDIENTE - LISTA I',
    'INDEPENDIENTES LISTA J': 'INDEPENDIENTE - LISTA J',
    'INDEPENDIENTES LISTA K': 'INDEPENDIENTE - LISTA K',
    'IZQUIERDA CIUDADANA DE CHILE': 'IZQUIERDA CIUDADANA DE CHILE',
    'IZQUIERDA CRISTIANA DE CHILE': 'IZQUIERDA CRISTIANA',
    'LIBERAL': 'PARTIDO LIBERAL',
    'LIBERAL DE CHILE': 'PARTIDO LIBERAL DE CHILE',
    'LOS VERDES': 'LOS VERDES',
    'MAS REGION': 'MAS REGION',
    'MAS REGION ': 'MAS REGION',
    'MOVIMIENTO AMPLIO SOCIAL': 'MOVIMIENTO AMPLIO SOCIAL',
    'MOVIMIENTO DE ACCION POPULAR U': 'MOVIMIENTO DE ACCION POPULAR U',
    'MOVIMIENTO ECOLOGISTA': 'MOVIMIENTO ECOLOGISTA',
    'MOVIMIENTO INDEPENDIENTE REGIONALISTA AGRARIO Y SOCIAL ': 'MOVIMIENTO INDEPENDIENTE REGIONALISTA AGRARIO Y SOCIAL',
    'NACIONAL': 'PARTIDO NACIONAL',
    'NUEVA ALIANZA POPULAR': 'NUEVA ALIANZA POPULAR',
    'P.A.I.S.': 'PARTIDO PAIS',
    'PAIS': 'PARTIDO PAIS',
    'PARTIDO COMUNISTA DE CHILE': 'PARTIDO COMUNISTA DE CHILE',
    'PARTIDO COMUNISTA DE CHILE ': 'PARTIDO COMUNISTA DE CHILE',
    'PARTIDO DE IZQUIERDA CIUDADANA DE CHILE': 'IZQUIERDA CIUDADANA DE CHILE',
    'PARTIDO DE TRABAJADORES REVOLUCIONARIOS': 'PARTIDO DE TRABAJADORES REVOLUCIONARIOS',
    'PARTIDO DEMOCRATA CRISTIANO': 'DEMOCRACIA CRISTIANA',
    'PARTIDO ECOLOGISTA VERDE': 'PARTIDO ECOLOGISTA VERDE',
    'PARTIDO ECOLOGISTA VERDE ': 'PARTIDO ECOLOGISTA VERDE',
    'PARTIDO ECOLOGISTA VERDE DEL NORTE': 'PARTIDO ECOLOGISTA VERDE DEL NORTE',
    'PARTIDO EVOLUCION POLITICA ': 'EVOLUCION POLITICA',
    'PARTIDO FRENTE POPULAR ': 'FRENTE POPULAR',
    'PARTIDO HUMANISTA': 'PARTIDO HUMANISTA',
    'PARTIDO IGUALDAD': 'PARTIDO IGUALDAD',
    'PARTIDO IZQUIERDA CIUDADANA DE CHILE': 'IZQUIERDA CIUDADANA DE CHILE',
    'PARTIDO LIBERAL DE CHILE': 'PARTIDO LIBERAL DE CHILE',
    'PARTIDO POR LA DEMOCRACIA': 'PPD',
    'PARTIDO PROGRESISTA': 'PARTIDO PROGRESISTA',
    'PARTIDO PROGRESISTA ': 'PARTIDO PROGRESISTA',
    'PARTIDO RADICAL SOCIALDEMOCRATA': 'PARTIDO RADICAL SOCIALDEMOCRATA',
    'PARTIDO REGIONALISTA DE LOS INDEPENDIENTES': 'PARTIDO REGIONALISTA DE LOS INDEPENDIENTES',
    'PARTIDO REGIONALISTA DE MAGALLANES ': 'PARTIDO REGIONALISTA DE MAGALLANES',
    'PARTIDO REGIONALISTA INDEPENDIENTE': 'PARTIDO REGIONALISTA INDEPENDIENTE',
    'PARTIDO SOCIALISTA DE CHILE': 'PARTIDO SOCIALISTA',
    'PARTIDO SOCIALISTA DE CHILE ': 'PARTIDO SOCIALISTA',
    'PODER': 'PODER',
    'POR LA DEMOCRACIA': 'PPD',
    'POR LA DEMOCRACIA ': 'PPD',
    'POR LA INTEGRACIÓN REGIONAL': 'PARTIDO POR LA INTEGRACION REGIONAL',
    'PROGRESISTA': 'PARTIDO PROGRESISTA',
    'RADICAL': 'PARTIDO RADICAL',
    'RADICAL DE CHILE': 'PARTIDO RADICAL',
    'RADICAL SOCIAL DEMOCRATA': 'PARTIDO RADICAL SOCIALDEMOCRATA',
    'RADICAL SOCIALDEMOCRATA': 'PARTIDO RADICAL SOCIALDEMOCRATA',
    'REGIONALISTA DE LOS INDEPENDIE': 'PARTIDO REGIONALISTA DE LOS INDEPENDIENTES',
    'REGIONALISTA DE LOS INDEPENDIENTES': 'PARTIDO REGIONALISTA DE LOS INDEPENDIENTES',
    'REGIONALISTA INDEPENDIENTE': 'PARTIDO REGIONALISTA INDEPENDIENTE',
    'RENOVACION NACIONAL': 'RENOVACION NACIONAL',
    'RENOVACION NACIONAL ': 'RENOVACION NACIONAL',
    'REVOLUCION DEMOCRATICA': 'REVOLUCION DEMOCRATICA',
    'SOCIALDEMOCRACIA CHILENA': 'SOCIALDEMOCRACIA CHILENA',
    'SOCIALISTA CHILENO': 'PARTIDO SOCIALISTA',
    'SOCIALISTA DE CHILE': 'PARTIDO SOCIALISTA',
    'SOMOS AYSEN': 'SOMOS AYSEN',
    'TODOS': 'TODOS',
    'UNIDOS RESULTA EN DEMOCRACIA': 'UNIDOS RESUELTA EN DEMOCRACIA',
    'UNION DE CENTRO CENTRO': 'UNION DE CENTRO CENTRO',
    'UNION DE CENTRO CENTRO PROGRES': 'UNION DE CENTRO CENTRO PROGRESISTA',
    'UNION DE CENTRO CENTRO PROGRESISTA': 'UNION DE CENTRO CENTRO PROGRESISTA',
    'UNION DEMOCRATA INDEPENDIENTE': 'UDI',
    'UNION DEMOCRATA INDEPENDIENTE ': 'UDI',
    'UNION PATRIOTICA': 'UNION PATRIOTICA',
    'WALLMAPUWEN': 'WALLMAPUWEN'
}

In [76]:
mapeo_partidos

{' ': 'NULOS/BLANCOS',
 '  ': 'NULOS/BLANCOS',
 'ALIANZA HUMANISTA VERDE': 'ALIANZA HUMANISTA VERDE',
 'ALIANZA HUMANISTA-VERDE': 'ALIANZA HUMANISTA VERDE',
 'ALIANZA NACIONAL DE LOS INDEPE': 'ALIANZA NACIONAL DE LOS INDEPENDIENTES',
 'ALIANZA NACIONAL DE LOS INDEPENDIENTES': 'ALIANZA NACIONAL DE LOS INDEPENDIENTES',
 'AMPLITUD': 'AMPLITUD',
 'AVANZADA NACIONAL': 'AVANZADA NACIONAL',
 'CANDIDATURA INDEPENDIENTE': 'INDEPENDIENTE',
 'CHILEPRIMERO': 'CHILEPRIMERO',
 'CIUDADANOS': 'CIUDADANOS',
 'COMUNISTA DE CHILE': 'PARTIDO COMUNISTA DE CHILE',
 'DE ACCION REGIONALISTA DE CHIL': 'PARTIDO DE ACCION REGIONALISTA DE CHILE',
 'DE ACCION REGIONALISTA DE CHILE': 'PARTIDO DE ACCION REGIONALISTA DE CHILE',
 'DE TRABAJADORES REVOLUCIONARIOS': 'PARTIDO DE TRABAJADORES REVOLUCIONARIOS',
 'DEL SUR': 'PARTIDO DEL SUR',
 'DEMOCRACIA RADICAL': 'DEMOCRACIA RADICAL',
 'DEMOCRACIA REGIONAL PATAGONICA': 'DEMOCRACIA REGIONAL PATAGONICA',
 'DEMOCRACIA REGIONAL PATAGONICA ': 'DEMOCRACIA REGIONAL PATAGONICA',


In [77]:
base_total["partido_2"] = base_total.partido.map(mapeo_partidos)

In [78]:
base_total.groupby(["partido", "partido_2"])["votos"].groups.keys()

dict_keys([(' ', 'NULOS/BLANCOS'), ('  ', 'NULOS/BLANCOS'), ('ALIANZA HUMANISTA VERDE', 'ALIANZA HUMANISTA VERDE'), ('ALIANZA HUMANISTA-VERDE', 'ALIANZA HUMANISTA VERDE'), ('ALIANZA NACIONAL DE LOS INDEPE', 'ALIANZA NACIONAL DE LOS INDEPENDIENTES'), ('ALIANZA NACIONAL DE LOS INDEPENDIENTES', 'ALIANZA NACIONAL DE LOS INDEPENDIENTES'), ('AMPLITUD', 'AMPLITUD'), ('AVANZADA NACIONAL', 'AVANZADA NACIONAL'), ('CANDIDATURA INDEPENDIENTE', 'INDEPENDIENTE'), ('CHILEPRIMERO', 'CHILEPRIMERO'), ('CIUDADANOS', 'CIUDADANOS'), ('COMUNISTA DE CHILE', 'PARTIDO COMUNISTA DE CHILE'), ('DE ACCION REGIONALISTA DE CHIL', 'PARTIDO DE ACCION REGIONALISTA DE CHILE'), ('DE ACCION REGIONALISTA DE CHILE', 'PARTIDO DE ACCION REGIONALISTA DE CHILE'), ('DE TRABAJADORES REVOLUCIONARIOS', 'PARTIDO DE TRABAJADORES REVOLUCIONARIOS'), ('DEL SUR', 'PARTIDO DEL SUR'), ('DEMOCRACIA RADICAL', 'DEMOCRACIA RADICAL'), ('DEMOCRACIA REGIONAL PATAGONICA', 'DEMOCRACIA REGIONAL PATAGONICA'), ('DEMOCRACIA REGIONAL PATAGONICA ', 'DEMO

Pareciera ser que los partidos ahora están mejor mapeados. Los independientes apoyados por un partido quedan como `INDEPENDIENTE - <PARTIDO>`, lo que aumenta la cantidad de partidos diferentes pero permite un análisis algo más acabado. Para ello, también, voy a crear una columna nueva, que indique el apoyo al independiente. Así, quedará el dataset con una columna `partido` y una columna `apoyo`.

In [79]:
base_total[base_total["partido"].isnull()].groupby(["candidato"]).groups.keys()

dict_keys([])

Primero, un chequeo pequeño de la cantidad de votos por partido, en una tabla pivote.

In [80]:
pd.pivot_table(data = base_total,
               values = "votos",
               index = ["partido_2", 'periodo'],
               aggfunc = np.sum,
               margins = False)

votos
partido_2                              periodo         
ALIANZA HUMANISTA VERDE                1992-1996  52481
                                       1994-1998  67733
                                       1994-2000  81814
                                       1994-2002   8528
ALIANZA NACIONAL DE LOS INDEPENDIENTES 2004-2008  45422
...                                                 ...
UNION DE CENTRO CENTRO PROGRESISTA     1998-2006  18023
UNION PATRIOTICA                       2016-2020  17105
                                       2018-2020  58341
                                       2018-2022  56601
WALLMAPUWEN                            2016-2020   3405

[382 rows x 1 columns]

Ahora, la columna `apoyo`.

In [81]:
def get_support(value):
    if value.find("-") > 0:
        value_1 = "INDEPENDIENTE"
        value_2 = value[value.find("-") + 1:]
    else:
        value_1 = value
        value_2 = None
    return value_1, value_2

In [82]:
base_total["partido_3"] = base_total.apply(lambda row: get_support(row.partido_2)[0], axis = 1)
base_total["apoyo"] = base_total.apply(lambda row: get_support(row.partido_2)[1], axis = 1)

In [83]:
base_total.head(n = 20)

index                   candidato       cargo     comuna  distrito lista  \
0       0  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE      ARICA       NaN   NaN   
1       1                HERNAN BUCHI  PRESIDENTE      ARICA       NaN   NaN   
2       2             PATRICIO AYLWIN  PRESIDENTE      ARICA       NaN   NaN   
3       3                VOTOS NULOS   PRESIDENTE      ARICA       NaN   NaN   
4       4            VOTOS EN BLANCO   PRESIDENTE      ARICA       NaN   NaN   
5       5  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE      ARICA       NaN   NaN   
6       6                HERNAN BUCHI  PRESIDENTE      ARICA       NaN   NaN   
7       7             PATRICIO AYLWIN  PRESIDENTE      ARICA       NaN   NaN   
8       8                VOTOS NULOS   PRESIDENTE      ARICA       NaN   NaN   
9       9            VOTOS EN BLANCO   PRESIDENTE      ARICA       NaN   NaN   
10     10  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE  CAMARONES       NaN   NaN   
11     11                HERNAN BUCHI  PRESIDENTE  CAMARONES       NaN   NaN   
12     12             PATRICIO AYLWIN  PRESIDENTE  CAMARONES       NaN   NaN   
13     13                VOTOS NULOS   PRESIDENTE  CAMARONES       NaN   NaN   
14     14            VOTOS EN BLANCO   PRESIDENTE  CAMARONES       NaN   NaN   
15     15  FRANCISCO JAVIER ERRAZURIZ  PRESIDENTE  CAMARONES       NaN   NaN   
16     16                HERNAN BUCHI  PRESIDENTE  CAMARONES       NaN   NaN   
17     17             PATRICIO AYLWIN  PRESIDENTE  CAMARONES       NaN   NaN   
18     18                VOTOS NULOS   PRESIDENTE  CAMARONES       NaN   NaN   
19     19            VOTOS EN BLANCO   PRESIDENTE  CAMARONES       NaN   NaN   

   circunscripcion provincia              partido    periodo       region  \
0              NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
1              NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
2              NaN     ARICA  DEMOCRATA CRISTIANO  1990-1994  DE TARAPACA   
3              NaN     ARICA                       1990-1994  DE TARAPACA   
4              NaN     ARICA                       1990-1994  DE TARAPACA   
5              NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
6              NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
7              NaN     ARICA  DEMOCRATA CRISTIANO  1990-1994  DE TARAPACA   
8              NaN     ARICA                       1990-1994  DE TARAPACA   
9              NaN     ARICA                       1990-1994  DE TARAPACA   
10             NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
11             NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
12             NaN     ARICA  DEMOCRATA CRISTIANO  1990-1994  DE TARAPACA   
13             NaN     ARICA                       1990-1994  DE TARAPACA   
14             NaN     ARICA                       1990-1994  DE TARAPACA   
15             NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
16             NaN     ARICA        INDEPENDIENTE  1990-1994  DE TARAPACA   
17             NaN     ARICA  DEMOCRATA CRISTIANO  1990-1994  DE TARAPACA   
18             NaN     ARICA                       1990-1994  DE TARAPACA   
19             NaN     ARICA                       1990-1994  DE TARAPACA   

      sexo tipo_eleccion tipo_voto_presidencial  votos electo       date  \
0   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN   7065     no 1989-12-14   
1   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  11549     no 1989-12-14   
2   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  23230     si 1989-12-14   
3      NaN  PRESIDENCIAL         UNICA VOTACIÓN    561     no 1989-12-14   
4      NaN  PRESIDENCIAL         UNICA VOTACIÓN    392     no 1989-12-14   
5   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN   7916     no 1989-12-14   
6   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  14237     no 1989-12-14   
7   HOMBRE  PRESIDENCIAL         UNICA VOTACIÓN  21326     si 1989-12-14   
8      NaN  PRESIDENC

In [84]:
base_total.shape[0] == total_records

True

In [85]:
base_total[base_total.partido_2.str.contains("-")].head()

index                         candidato    cargo comuna  distrito  \
96902  30996      PIZARRO BOSSAY PABLO ANTONIO  SENADOR  ARICA       1.0   
96903  30997  ZAPATA VELASQUEZ SANDRA CECILIA   SENADOR  ARICA       1.0   
96908  31002   URRUTIA CARDENAS SALVADOR PEDRO  SENADOR  ARICA       1.0   
96909  31003    PARRA CORREA TRINIDAD VICTORIA  SENADOR  ARICA       1.0   
96911  31005          BARBOSA BARRIOS RODOLFO   SENADOR  ARICA       1.0   

      lista circunscripcion provincia                            partido  \
96902     B           ARICA       NaN          INDEPENDIENTE PROGRESISTA   
96903     B           ARICA       NaN          INDEPENDIENTE PROGRESISTA   
96908     N           ARICA       NaN    INDEPENDIENTE POR LA DEMOCRACIA   
96909     O           ARICA       NaN  INDEPENDIENTE DEMOCRATA CRISTIANO   
96911     P           ARICA       NaN  INDEPENDIENTE RENOVACION NACIONAL   

         periodo              region    sexo  tipo_eleccion  \
96902  2018-2020  ARICA Y PARINACOTA  HOMBRE  PARLAMENTARIA   
96903  2018-2020  ARICA Y PARINACOTA   MUJER  PARLAMENTARIA   
96908  2018-2020  ARICA Y PARINACOTA  HOMBRE  PARLAMENTARIA   
96909  2018-2020  ARICA Y PARINACOTA   MUJER  PARLAMENTARIA   
96911  2018-2020  ARICA Y PARINACOTA  HOMBRE  PARLAMENTARIA   

      tipo_voto_presidencial  votos electo       date  \
96902                    NaN    426     si 2017-11-19   
96903                    NaN    327     si 2017-11-19   
96908                    NaN   3295     si 2017-11-19   
96909                    NaN    336     si 2017-11-19   
96911                    NaN   3518     si 2017-11-19   

                                  partido_2      partido_3  \
96902   INDEPENDIENTE - PARTIDO PROGRESISTA  INDEPENDIENTE   
96903   INDEPENDIENTE - PARTIDO PROGRESISTA  INDEPENDIENTE   
96908                   INDEPENDIENTE - PPD  INDEPENDIENTE   
96909  INDEPENDIENTE - DEMOCRACIA CRISTIANA  INDEPENDIENTE   
96911   INDEPENDIENTE - RENOVACION NACIONAL  INDEPENDIENTE   

                       apoyo  
96902    PARTIDO PROGRESISTA  
96903    PARTIDO PROGRESISTA  
96908                    PPD  
96909   DEMOCRACIA CRISTIANA  
96911    RENOVACION NACIONAL

Vamos a ver la estructura actual del DataFrame y limpiar de columnas innecesarias.

In [86]:
base_total.columns.tolist()

['index',
 'candidato',
 'cargo',
 'comuna',
 'distrito',
 'lista',
 'circunscripcion',
 'provincia',
 'partido',
 'periodo',
 'region',
 'sexo',
 'tipo_eleccion',
 'tipo_voto_presidencial',
 'votos',
 'electo',
 'date',
 'partido_2',
 'partido_3',
 'apoyo']

In [87]:
base_total.drop(columns = ["partido", "partido_2", "tipo_voto_presidencial"], inplace = True)
base_total.rename(columns = {"partido_3": "partido"}, inplace = True)

In [88]:
base_total.columns.tolist()

['index',
 'candidato',
 'cargo',
 'comuna',
 'distrito',
 'lista',
 'circunscripcion',
 'provincia',
 'periodo',
 'region',
 'sexo',
 'tipo_eleccion',
 'votos',
 'electo',
 'date',
 'partido',
 'apoyo']

In [89]:
base_total.votos.sum() == total_votes

True

In [90]:
base_total.shape[0] == total_records

True

Ahora revisaremos el campo `lista`.

In [91]:
base_total.lista.unique()

array([nan, 'A', 'B', 'CI', 'E', 'G', 'D', 'F', 'C', 'H', 'J', 'L', 'I',
       'N', 'O', 'P', 'R', 'K', 'M', ' ', 'T', 'Q', 'S'], dtype=object)

In [92]:
base_total[(base_total.lista.notnull())]["tipo_eleccion"].unique()

array(['PARLAMENTARIA', 'MUNICIPALES', 'CORE'], dtype=object)

In [93]:
base_total[(base_total.lista.notnull())]["periodo"].unique()

array(['1990-1994', '1994-2002', '1998-2006', '2002-2010', '2006-2014',
       '2010-2018', '2014-2022', '2018-2020', '1994-1998', '1998-2002',
       '2002-2006', '2006-2010', '2010-2014', '2014-2018', '2004-2008',
       '2008-2012', '2012-2016', '2016-2020', '1992-1996', '1996-2000',
       '2000-2004', '2018-2022'], dtype=object)

Me da la impresión de que `lista` no es muy útil ni tiene mucha información, así que lo eliminamos también.

In [94]:
base_total.drop(columns = ["lista"], inplace = True)

La columna `sexo` tampoco es relevante, así que se elimina.

In [95]:
base_total.sexo.unique()

array(['HOMBRE', nan, 'MUJER', '  ', ' '], dtype=object)

In [96]:
base_total.drop(columns = ["sexo"], inplace = True)

Finalmente, queda limpiar los votos en blanco y votos nulos.

In [97]:
base_total[(base_total.candidato.str.contains("VOTO"))].candidato.unique()

array(['VOTOS NULOS ', 'VOTOS EN BLANCO ', 'VOTOS NULOS',
       'VOTOS EN BLANCO'], dtype=object)

In [98]:
base_total[(base_total.candidato.str.contains("VOTOS EN BLANCO"))].candidato.unique()

array(['VOTOS EN BLANCO ', 'VOTOS EN BLANCO'], dtype=object)

In [99]:
base_total[(base_total.candidato.str.contains("VOTOS NULOS"))].candidato.unique()

array(['VOTOS NULOS ', 'VOTOS NULOS'], dtype=object)

In [100]:
base_total.loc[(base_total.candidato.str.contains("VOTOS EN BLANCO")), "candidato"] = "BLANCOS"
base_total.loc[(base_total.candidato.str.contains("VOTOS NULOS")), "candidato"] = "NULOS"

In [101]:
base_total[(base_total.partido.str.contains("BLANCOS"))].candidato.unique()

array(['NULOS', 'BLANCOS'], dtype=object)

In [102]:
base_total[(base_total.candidato.str.contains("BLANCO"))].candidato.unique()

array(['BLANCOS', 'BLANCO URZUA DAVID', 'BLANCO OVALLE NESTOR ALIRO',
       'HERRERA BLANCO PATRICIO', 'BLANCO HADI CARMEN',
       'ARRIAGADA BLANCO CARLOS', 'CARMEN BLANCO HADI',
       'CHOQUE BLANCO MAGDALENA', 'ARIAS BLANCO JUAN ALBERTO',
       'UNDURRAGA CASTELBLANCO MARIO', 'BLANCO WATSON RAUL',
       'CHAMORRO BLANCO JUAN', 'BLANCO MEJIAS MIGUEL',
       'BLANCO VALENCIA MARIA DEL PILAR',
       'BLANCO CHIHUAILAF LUIS ALBERTO', 'BLANCO LEMUNAO VICENTE',
       'LLANCALEO BLANCO MARIO', 'JOSE LUIS BLANCO VASQUEZ',
       'MAURICIO ALEJANDRO RIVERA BLANCO', 'DAISY PAOLA GUTIERREZ BLANCO',
       'JUAN ENRIQUE CHAMORRO BLANCO',
       'MARGARITA DEL CARMEN BLANCO SILVA',
       'MARIO UNDURRAGA CASTELBLANCO', 'PATRICIO HERRERA BLANCO',
       'MARIO LLANCALEO BLANCO', 'GUILLERMO BLANCO ANTINAO',
       'LUIS BLANCO AREVALO', 'EDUARDO EBNER BLANCO',
       'DAVID BLANCO URZUA', 'VIRGINIA CHOQUE BLANCO',
       'JUAN PABLO BLANCO BLANCO', 'LEONOR EMA MUÑOZ BLANCO',
       'ANDRE

In [103]:
base_total[(base_total.candidato.str.contains("NULO"))].candidato.unique()

array(['NULOS'], dtype=object)

In [104]:
base_total[(base_total.candidato.str.contains("VOTO"))].candidato.unique()

array([], dtype=object)

Se revisarán ahora las variables geográficas: `distrito`, `comuna`, `circunscripcion`, `region` y `provincia`.

Parto con `region`.

In [105]:
base_total.region.unique()

array(['DE TARAPACA', 'DE ANTOFAGASTA', 'DE ATACAMA', 'DE COQUIMBO',
       'DE VALPARAISO', "DEL LIBERTADOR BDO. O'HIGGINS", 'DEL MAULE',
       'DEL BIOBIO', 'DE LA ARAUCANIA', 'DE LOS LAGOS',
       'AISEN DEL GRAL. CARLOS IBAÑEZ', 'DE MAGALLANES Y ANTARTICA CH.',
       'METROPOLITANA DE SANTIAGO', 'DE LOS RIOS', 'ARICA Y PARINACOTA',
       'DEL LIBERTADOR BDO. OHIGGINS',
       'METROPOLITANA DE SANTIAGO               ', nan], dtype=object)

In [106]:
base_total.region.str.strip().unique()

array(['DE TARAPACA', 'DE ANTOFAGASTA', 'DE ATACAMA', 'DE COQUIMBO',
       'DE VALPARAISO', "DEL LIBERTADOR BDO. O'HIGGINS", 'DEL MAULE',
       'DEL BIOBIO', 'DE LA ARAUCANIA', 'DE LOS LAGOS',
       'AISEN DEL GRAL. CARLOS IBAÑEZ', 'DE MAGALLANES Y ANTARTICA CH.',
       'METROPOLITANA DE SANTIAGO', 'DE LOS RIOS', 'ARICA Y PARINACOTA',
       'DEL LIBERTADOR BDO. OHIGGINS', nan], dtype=object)

In [107]:
base_total["region"] = base_total.region.str.strip()

In [108]:
base_total.region.unique()

array(['DE TARAPACA', 'DE ANTOFAGASTA', 'DE ATACAMA', 'DE COQUIMBO',
       'DE VALPARAISO', "DEL LIBERTADOR BDO. O'HIGGINS", 'DEL MAULE',
       'DEL BIOBIO', 'DE LA ARAUCANIA', 'DE LOS LAGOS',
       'AISEN DEL GRAL. CARLOS IBAÑEZ', 'DE MAGALLANES Y ANTARTICA CH.',
       'METROPOLITANA DE SANTIAGO', 'DE LOS RIOS', 'ARICA Y PARINACOTA',
       'DEL LIBERTADOR BDO. OHIGGINS', nan], dtype=object)

In [109]:
mapeo_regiones = {
    "DE TARAPACA": "TARAPACA",
    "DE ANTOFAGASTA": "ANTOFAGASTA",
    "DE ATACAMA": "ATACAMA",
    "DE COQUIMBO": "COQUIMBO",
    "DE VALPARAISO": "VALPARAISO",
    "DEL LIBERTADOR BDO. O'HIGGINS": "O'HIGGINS",
    "DEL MAULE": "MAULE",
    "DEL BIOBIO": "BIOBIO",
    "DE LA ARAUCANIA": "ARAUCANIA",
    "DE LOS LAGOS": "LOS LAGOS",
    "AISEN DEL GRAL. CARLOS IBAÑEZ": "AYSEN",
    "DE MAGALLANES Y ANTARTICA CH.": "MAGALLANES",
    "METROPOLITANA DE SANTIAGO": "METROPOLITANA",
    "DE LOS RIOS": "LOS RIOS",
    "ARICA Y PARINACOTA": "ARICA",
    "DEL LIBERTADOR BDO OHIGGINS": "O'HIGGINS"
}

In [110]:
base_total["region_2"] = base_total.region.map(mapeo_regiones)

In [111]:
base_total.region_2.unique()

array(['TARAPACA', 'ANTOFAGASTA', 'ATACAMA', 'COQUIMBO', 'VALPARAISO',
       "O'HIGGINS", 'MAULE', 'BIOBIO', 'ARAUCANIA', 'LOS LAGOS', 'AYSEN',
       'MAGALLANES', 'METROPOLITANA', 'LOS RIOS', 'ARICA', nan],
      dtype=object)

In [112]:
base_total[(base_total.region_2.isnull())]

index                       candidato       cargo   comuna  distrito  \
48331  48331            ALFREDO SFEIR YOUNIS  PRESIDENTE  CHEPICA       NaN   
48332  48332           EVELYN MATTHEI FORNET  PRESIDENTE  CHEPICA       NaN   
48333  48333    FRANCO ALDO PARISI FERNANDEZ  PRESIDENTE  CHEPICA       NaN   
48334  48334             MARCEL CLAUDE REYES  PRESIDENTE  CHEPICA       NaN   
48335  48335  MARCO ENRIQUEZ-OMINAMI GUMUCIO  PRESIDENTE  CHEPICA       NaN   
...      ...                             ...         ...      ...       ...   
65901   1535                         BLANCOS  PRESIDENTE      NaN       NaN   
65902   1536     ALEJANDRO  GUILLIER ALVAREZ  PRESIDENTE      NaN       NaN   
65903   1537      SEBASTIAN PIÑERA ECHENIQUE  PRESIDENTE      NaN       NaN   
65904   1538                           NULOS  PRESIDENTE      NaN       NaN   
65905   1539                         BLANCOS  PRESIDENTE      NaN       NaN   

      circunscripcion  provincia    periodo                        region  \
48331             NaN  COLCHAGUA  2014-2018  DEL LIBERTADOR BDO. OHIGGINS   
48332             NaN  COLCHAGUA  2014-2018  DEL LIBERTADOR BDO. OHIGGINS   
48333             NaN  COLCHAGUA  2014-2018  DEL LIBERTADOR BDO. OHIGGINS   
48334             NaN  COLCHAGUA  2014-2018  DEL LIBERTADOR BDO. OHIGGINS   
48335             NaN  COLCHAGUA  2014-2018  DEL LIBERTADOR BDO. OHIGGINS   
...               ...        ...        ...                           ...   
65901             NaN        NaN  2018-2022                           NaN   
65902             NaN        NaN  2018-2022                           NaN   
65903             NaN        NaN  2018-2022                           NaN   
65904             NaN        NaN  2018-2022                           NaN   
65905             NaN        NaN  2018-2022                           NaN   

      tipo_eleccion  votos electo       date                   partido apoyo  \
48331  PRESIDENCIAL     29     no 2013-11-17  PARTIDO ECOLOGISTA VERDE  None   
48332  PRESIDENCIAL   1991     si 2013-11-17                       UDI  None   
48333  PRESIDENCIAL    381     no 2013-11-17             INDEPENDIENTE  None   
48334  PRESIDENCIAL     44     no 2013-11-17         PARTIDO HUMANISTA  None   
48335  PRESIDENCIAL    396     no 2013-11-17       PARTIDO PROGRESISTA  None   
...             ...    ...    ...        ...                       ...   ...   
65901  PRESIDENCIAL      0     no 2017-12-17             NULOS/BLANCOS  None   
65902  PRESIDENCIAL     63     si 2017-12-17             INDEPENDIENTE  None   
65903  PRESIDENCIAL    141     si 2017-12-17             INDEPENDIENTE  None   
65904  PRESIDENCIAL      0     no 2017-12-17             NULOS/BLANCOS  None   
65905  PRESIDENCIAL      0     no 2017-12-17             NULOS/BLANCOS  None   

      region_2  
48331      NaN  
48332      NaN  
48333      NaN  
48334      NaN  
48335      NaN  
...        ...  
65901      NaN  
65902      NaN  
65903      NaN  
65904      NaN  
65905      NaN  

[2200 rows x 16 columns]

In [113]:
base_total[(base_total.region_2.isnull())].groupby(["comuna", "distrito", "provincia"]).groups.keys()

dict_keys([('CHEPICA', nan, 'COLCHAGUA'), ('CHIMBARONGO', nan, 'COLCHAGUA'), ('CODEGUA', nan, 'CACHAPOAL'), ('COINCO', nan, 'CACHAPOAL'), ('COLTAUCO', nan, 'CACHAPOAL'), ('DOÑIHUE', nan, 'CACHAPOAL'), ('GRANEROS', nan, 'CACHAPOAL'), ('LA ESTRELLA', nan, 'CARDENAL CARO'), ('LAS CABRAS', nan, 'CACHAPOAL'), ('LITUECHE', nan, 'CARDENAL CARO'), ('LOLOL', nan, 'COLCHAGUA'), ('MACHALI', nan, 'CACHAPOAL'), ('MALLOA', nan, 'CACHAPOAL'), ('MARCHIGUE', nan, 'CARDENAL CARO'), ('MOSTAZAL', nan, 'CACHAPOAL'), ('NANCAGUA', nan, 'COLCHAGUA'), ('NAVIDAD', nan, 'CARDENAL CARO'), ('OLIVAR', nan, 'CACHAPOAL'), ('PALMILLA', nan, 'COLCHAGUA'), ('PAREDONES', nan, 'CARDENAL CARO'), ('PERALILLO', nan, 'COLCHAGUA'), ('PEUMO', nan, 'CACHAPOAL'), ('PICHIDEGUA', nan, 'CACHAPOAL'), ('PICHILEMU', nan, 'CARDENAL CARO'), ('PLACILLA', nan, 'COLCHAGUA'), ('PUMANQUE', nan, 'COLCHAGUA'), ('QUINTA DE TILCOCO', nan, 'CACHAPOAL'), ('RANCAGUA', nan, 'CACHAPOAL'), ('RENGO', nan, 'CACHAPOAL'), ('REQUINOA', nan, 'CACHAPOAL'), ('

In [114]:
base_total[(base_total.region_2.isnull()) & \
           (base_total.comuna.isnull()) & \
           (base_total.distrito.isnull()) & \
           (base_total.provincia.isnull())].tipo_eleccion.unique()

array(['PRESIDENCIAL'], dtype=object)

In [115]:
base_total[(base_total.tipo_eleccion == "PRESIDENCIAL")].region_2.unique()

array(['TARAPACA', 'ANTOFAGASTA', 'ATACAMA', 'COQUIMBO', 'VALPARAISO',
       "O'HIGGINS", 'MAULE', 'BIOBIO', 'ARAUCANIA', 'LOS LAGOS', 'AYSEN',
       'MAGALLANES', 'METROPOLITANA', 'LOS RIOS', 'ARICA', nan],
      dtype=object)

Hay instancias donde `region` es `NULL`, y sospecho que los valores de votación de `tipo_eleccion` == `PRESIDENCIAL` que vengan del extranjero tendrán región `NULL` además de `comuna`, `distrito` y `provincia` nulas, lo que se puede validar con la cantidad de votos.

Por otro lado, hay otras tuplas que tienen `region` nula pero tienen información de `comuna`, `distrito` o `provincia`, lo que permitirá limpiar dichos datos y asociarlos a una región en particular.

Verificando que los votos con `tipo_eleccion` == `PRESIDENCIAL` sean iguales a la cantidad de entrada...

In [116]:
tmp_df = base_total[(base_total.tipo_eleccion == "PRESIDENCIAL")].copy()
tmp_df["region_2"] = tmp_df.region_2.astype(str)
tmp_df.groupby(["region_2"])["votos"].sum()

region_2
ANTOFAGASTA       2300700
ARAUCANIA         4852700
ARICA              436336
ATACAMA           1303583
AYSEN              479360
BIOBIO           10793862
COQUIMBO          3184537
LOS LAGOS         4952342
LOS RIOS           969229
MAGALLANES         828016
MAULE             5426378
METROPOLITANA    33022056
O'HIGGINS         3971833
TARAPACA          1600330
VALPARAISO        9256047
nan                733126
Name: votos, dtype: int64

In [117]:
tmp_df[((base_total.region_2.isnull()) & \
        (base_total.comuna.isnull()) & \
        (base_total.distrito.isnull()) & \
        (base_total.provincia.isnull()))].groupby(["region_2"])["votos"].sum()

/home/jtapia/anaconda3/envs/data_analysis/lib/python3.6/site-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


region_2
nan    44626
Name: votos, dtype: int64

De una inspección al archivo de entrada, queda claro que la región faltante es `EXTERIOR` (la suma de votos en el archivo Excel de entrada es de 44,626). Entonces, le asignaremos dicha región para mantener la consistencia de los datos.

In [118]:
base_total.loc[(base_total.region_2.isnull()) & \
               (base_total.comuna.isnull()) & \
               (base_total.distrito.isnull()) & \
               (base_total.provincia.isnull()) & \
               (base_total.tipo_eleccion == "PRESIDENCIAL"), "region_2"] = "EXTERIOR"

In [119]:
tmp_df = base_total[(base_total.tipo_eleccion == "PRESIDENCIAL")].copy()
tmp_df["region_2"] = tmp_df.region_2.astype(str)
tmp_df.groupby(["region_2"])["votos"].sum()

region_2
ANTOFAGASTA       2300700
ARAUCANIA         4852700
ARICA              436336
ATACAMA           1303583
AYSEN              479360
BIOBIO           10793862
COQUIMBO          3184537
EXTERIOR            44626
LOS LAGOS         4952342
LOS RIOS           969229
MAGALLANES         828016
MAULE             5426378
METROPOLITANA    33022056
O'HIGGINS         3971833
TARAPACA          1600330
VALPARAISO        9256047
nan                688500
Name: votos, dtype: int64

In [120]:
base_total[base_total.region_2.isnull()].groupby(["tipo_eleccion", "periodo", "candidato"])["votos"].sum()

tipo_eleccion  periodo    candidato                     
PRESIDENCIAL   2014-2018  ALFREDO SFEIR YOUNIS                5494
                          BLANCOS                             4128
                          EVELYN MATTHEI FORNET             191379
                          FRANCO ALDO PARISI FERNANDEZ       35039
                          MARCEL CLAUDE REYES                 7466
                          MARCO ENRIQUEZ-OMINAMI GUMUCIO     31402
                          MICHELLE BACHELET JERIA           400712
                          NULOS                               7083
                          RICARDO ISRAEL ZIPPER               1710
                          ROXANA MIRANDA MENESES              3263
                          TOMAS JOCELYN-HOLT LETELIER          824
Name: votos, dtype: int64

In [121]:
base_total[base_total.region_2.isnull()].groupby(["provincia"]).groups.keys()

dict_keys(['CACHAPOAL', 'CARDENAL CARO', 'COLCHAGUA'])

In [122]:
base_total.loc[(base_total.region_2.isnull()), "region_2"] = "O'HIGGINS"

In [123]:
tmp_df = base_total[(base_total.tipo_eleccion == "PRESIDENCIAL")].copy()
tmp_df["region_2"] = tmp_df.region_2.astype(str)
tmp_df.groupby(["region_2"])["votos"].sum()

region_2
ANTOFAGASTA       2300700
ARAUCANIA         4852700
ARICA              436336
ATACAMA           1303583
AYSEN              479360
BIOBIO           10793862
COQUIMBO          3184537
EXTERIOR            44626
LOS LAGOS         4952342
LOS RIOS           969229
MAGALLANES         828016
MAULE             5426378
METROPOLITANA    33022056
O'HIGGINS         4660333
TARAPACA          1600330
VALPARAISO        9256047
Name: votos, dtype: int64

In [124]:
base_total[base_total.region_2.isnull()]["tipo_eleccion"].unique()

array([], dtype=object)

In [125]:
base_total.region_2.unique()

array(['TARAPACA', 'ANTOFAGASTA', 'ATACAMA', 'COQUIMBO', 'VALPARAISO',
       "O'HIGGINS", 'MAULE', 'BIOBIO', 'ARAUCANIA', 'LOS LAGOS', 'AYSEN',
       'MAGALLANES', 'METROPOLITANA', 'LOS RIOS', 'ARICA', 'EXTERIOR'],
      dtype=object)

In [126]:
base_total.drop(columns = ["region"], inplace = True)
base_total.rename(columns = {"region_2": "region"}, inplace = True)

In [127]:
base_total.region.unique()

array(['TARAPACA', 'ANTOFAGASTA', 'ATACAMA', 'COQUIMBO', 'VALPARAISO',
       "O'HIGGINS", 'MAULE', 'BIOBIO', 'ARAUCANIA', 'LOS LAGOS', 'AYSEN',
       'MAGALLANES', 'METROPOLITANA', 'LOS RIOS', 'ARICA', 'EXTERIOR'],
      dtype=object)

Limpieza de entorno.

In [128]:
%whos

Variable             Type         Data/Info
-------------------------------------------
base_total           DataFrame            index            <...>374720 rows x 15 columns]
condiciones          Series       0         False\n1       <...>ngth: 374720, dtype: bool
condiciones_electo   list         n=4
get_support          function     <function get_support at 0x7f7d15370840>
mapeo_partidos       dict         n=131
mapeo_regiones       dict         n=16
mpl                  module       <module 'matplotlib' from<...>/matplotlib/__init__.py'>
np                   module       <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
pd                   module       <module 'pandas' from '/h<...>ages/pandas/__init__.py'>
resultados_electo    list         n=4
subprocess           module       <module 'subprocess' from<...>python3.6/subprocess.py'>
tmp                  list         n=8
tmp_df               DataFrame           index             <...>[65906 rows x 16 columns]
total_records 

In [129]:
del condiciones, condiciones_electo, mapeo_partidos, mapeo_regiones, resultados_electo, tmp, tmp_df

Seguimos con `provincia`.

In [130]:
base_total[base_total.provincia.isnull()]

index                      candidato               cargo  \
64366       0    ALEJANDRO  GUILLIER ALVAREZ          PRESIDENTE   
64367       1        ALEJANDRO NAVARRO BRAIN          PRESIDENTE   
64368       2          BEATRIZ SANCHEZ MUÑOZ          PRESIDENTE   
64369       3         CAROLINA GOIC BOROEVIC          PRESIDENTE   
64370       4        EDUARDO ARTES BRICHETTI          PRESIDENTE   
...       ...                            ...                 ...   
374715  33765  SOTO ESPAÑA TOLENTINO SEGUNDO  CONSEJERO REGIONAL   
374716  33766            VALDES LUFI GONZALO  CONSEJERO REGIONAL   
374717  33767          WARNER ALVAREZ YAMMY   CONSEJERO REGIONAL   
374718  33768                          NULOS  CONSEJERO REGIONAL   
374719  33769                        BLANCOS  CONSEJERO REGIONAL   

                  comuna  distrito                 circunscripcion provincia  \
64366                NaN       NaN                             NaN       NaN   
64367                NaN       NaN                             NaN       NaN   
64368                NaN       NaN                             NaN       NaN   
64369                NaN       NaN                             NaN       NaN   
64370                NaN       NaN                             NaN       NaN   
...                  ...       ...                             ...       ...   
374715  TORRES DEL PAINE      28.0  TORRES DEL PAINE (C. CASTILLO)       NaN   
374716  TORRES DEL PAINE      28.0  TORRES DEL PAINE (C. CASTILLO)       NaN   
374717  TORRES DEL PAINE      28.0  TORRES DEL PAINE (C. CASTILLO)       NaN   
374718  TORRES DEL PAINE      28.0  TORRES DEL PAINE (C. CASTILLO)       NaN   
374719  TORRES DEL PAINE      28.0  TORRES DEL PAINE (C. CASTILLO)       NaN   

          periodo tipo_eleccion  votos electo       date  \
64366   2018-2022  PRESIDENCIAL    101     si 2017-11-19   
64367   2018-2022  PRESIDENCIAL      0     no 2017-11-19   
64368   2018-2022  PRESIDENCIAL    312     no 2017-11-19   
64369   2018-2022  PRESIDENCIAL     29     no 2017-11-19   
64370   2018-2022  PRESIDENCIAL      5     no 2017-11-19   
...           ...           ...    ...    ...        ...   
374715  2018-2022          CORE    152     si 2017-11-19   
374716  2018-2022          CORE      7     si 2017-11-19   
374717  2018-2022          CORE     59     si 2017-11-19   
374718  2018-2022          CORE     18     no 2017-11-19   
374719  2018-2022          CORE     28     no 2017-11-19   

                     partido apoyo      region  
64366          INDEPENDIENTE  None    EXTERIOR  
64367           PARTIDO PAIS  None    EXTERIOR  
64368          INDEPENDIENTE  None    EXTERIOR  
64369   DEMOCRACIA CRISTIANA  None    EXTERIOR  
64370       UNION PATRIOTICA  None    EXTERIOR  
...                      ...   ...         ...  
374715         INDEPENDIENTE  None  MAGALLANES  
374716    EVOLUCION POLITICA  None  MAGALLANES  
374717                   UDI  None  MAGALLANES  
374718         NULOS/BLANCOS  None  MAGALLANES  
374719         NULOS/BLANCOS  None  MAGALLANES  

[310354 rows x 15 columns]

In [131]:
base_total[(base_total.provincia.isnull())].region.unique()

array(['EXTERIOR', 'TARAPACA', 'ANTOFAGASTA', 'ATACAMA', 'COQUIMBO',
       'VALPARAISO', 'METROPOLITANA', "O'HIGGINS", 'MAULE', 'BIOBIO',
       'ARAUCANIA', 'LOS LAGOS', 'AYSEN', 'MAGALLANES', 'ARICA',
       'LOS RIOS'], dtype=object)

In [132]:
tmp_df = base_total.copy()
tmp_df["provincia"] = tmp_df.provincia.astype(str)
tmp_df[(tmp_df.provincia == "nan") & (tmp_df.tipo_eleccion != "PRESIDENCIAL")] \
    .groupby(["tipo_eleccion", "region", "provincia"])["votos"].count()

tipo_eleccion  region         provincia
CORE           ANTOFAGASTA    nan           1094
               ARAUCANIA      nan           3664
               ARICA          nan            846
               ATACAMA        nan            671
               AYSEN          nan            867
               BIOBIO         nan           5494
               COQUIMBO       nan           1789
               LOS LAGOS      nan           3971
               LOS RIOS       nan           2860
               MAGALLANES     nan            497
               MAULE          nan           3134
               METROPOLITANA  nan           3687
               O'HIGGINS      nan           2540
               TARAPACA       nan            555
               VALPARAISO     nan           2101
MUNICIPALES    ANTOFAGASTA    nan           4414
               ARAUCANIA      nan          15630
               ARICA          nan           1398
               ATACAMA        nan           3827
               AYSEN         

La base parece estar limpia. Ahora, se exporta a CSV para cargarla en un libro de Tableau y poder armar un libro público con que muestre resultados.

In [133]:
# write_file = "/mnt/c/Users/javie/Documents/Projects/Analytics/20191116-elecciones_chile/resultados.csv"
# base_total.to_csv(path_or_buf = write_file, index = False)